
<!doctype html>
<html lang="en" class="no-js">
  <head>
    
      <meta charset="utf-8">
      <meta name="viewport" content="width=device-width,initial-scale=1">
      
      
        <meta name="author" content="['Vadim I. Nazarov']">
      
      
        <link rel="canonical" href="https://immunarch.com/tutorials/single_cell/">
      
      
        <link rel="prev" href="../../howto/no_threads/">
      
      
        <link rel="next" href="../migration/">
      
      
      <link rel="icon" href="../../assets/images/favicon.png">
      <meta name="generator" content="mkdocs-1.6.1, mkdocs-material-9.6.22">
    
    
      
        <title>Comprehensive analysis of single-cell immune repertoire data using `immunarch` and `immundata` - immunomind.</title>
      
    
    
      <link rel="stylesheet" href="../../assets/stylesheets/main.84d31ad4.min.css">
      
        
        <link rel="stylesheet" href="../../assets/stylesheets/palette.06af60db.min.css">
      
      


    
    
      
    
    
      
        
        
        <link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>
        <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Roboto:300,300i,400,400i,700,700i%7CRoboto+Mono:400,400i,700,700i&display=fallback">
        <style>:root{--md-text-font:"Roboto";--md-code-font:"Roboto Mono"}</style>
      
    
    
      <link rel="stylesheet" href="../../stylesheets/extra.css">
    
    <script>__md_scope=new URL("../..",location),__md_hash=e=>[...e].reduce(((e,_)=>(e<<5)-e+_.charCodeAt(0)),0),__md_get=(e,_=localStorage,t=__md_scope)=>JSON.parse(_.getItem(t.pathname+"."+e)),__md_set=(e,_,t=localStorage,a=__md_scope)=>{try{t.setItem(a.pathname+"."+e,JSON.stringify(_))}catch(e){}}</script>
    
      
  


  
  

<script id="__analytics">function __md_analytics(){function e(){dataLayer.push(arguments)}window.dataLayer=window.dataLayer||[],e("js",new Date),e("config","G-G5JZNSNPH6"),document.addEventListener("DOMContentLoaded",(function(){document.forms.search&&document.forms.search.query.addEventListener("blur",(function(){this.value&&e("event","search",{search_term:this.value})}));document$.subscribe((function(){var t=document.forms.feedback;if(void 0!==t)for(var a of t.querySelectorAll("[type=submit]"))a.addEventListener("click",(function(a){a.preventDefault();var n=document.location.pathname,d=this.getAttribute("data-md-value");e("event","feedback",{page:n,data:d}),t.firstElementChild.disabled=!0;var r=t.querySelector(".md-feedback__note [data-md-value='"+d+"']");r&&(r.hidden=!1)})),t.hidden=!1})),location$.subscribe((function(t){e("config","G-G5JZNSNPH6",{page_path:t.pathname})}))}));var t=document.createElement("script");t.async=!0,t.src="https://www.googletagmanager.com/gtag/js?id=G-G5JZNSNPH6",document.getElementById("__analytics").insertAdjacentElement("afterEnd",t)}</script>
  
    <script>"undefined"!=typeof __md_analytics&&__md_analytics()</script>
  

    
    
    
  </head>
  
  
    
    
      
    
    
    
    
    <body dir="ltr" data-md-color-scheme="default" data-md-color-primary="blue-grey" data-md-color-accent="amber">
  
    
    <input class="md-toggle" data-md-toggle="drawer" type="checkbox" id="__drawer" autocomplete="off">
    <input class="md-toggle" data-md-toggle="search" type="checkbox" id="__search" autocomplete="off">
    <label class="md-overlay" for="__drawer"></label>
    <div data-md-component="skip">
      
        
        <a href="#introduction" class="md-skip">
          Skip to content
        </a>
      
    </div>
    <div data-md-component="announce">
      
    </div>
    
    
      

  

<header class="md-header md-header--shadow md-header--lifted" data-md-component="header">
  <nav class="md-header__inner md-grid" aria-label="Header">
    <a href="../.." title="immunomind." class="md-header__button md-logo" aria-label="immunomind." data-md-component="logo">
      
  <img src="../../assets/phoenix_cl2.png" alt="logo">

    </a>
    <label class="md-header__button md-icon" for="__drawer">
      
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M3 6h18v2H3zm0 5h18v2H3zm0 5h18v2H3z"/></svg>
    </label>
    <div class="md-header__title" data-md-component="header-title">
      <div class="md-header__ellipsis">
        <div class="md-header__topic">
          <span class="md-ellipsis">
            immunomind.
          </span>
        </div>
        <div class="md-header__topic" data-md-component="header-topic">
          <span class="md-ellipsis">
            
              Comprehensive analysis of single-cell immune repertoire data using `immunarch` and `immundata`
            
          </span>
        </div>
      </div>
    </div>
    
      
        <form class="md-header__option" data-md-component="palette">
  
    
    
    
    <input class="md-option" data-md-color-media="" data-md-color-scheme="default" data-md-color-primary="blue-grey" data-md-color-accent="amber"  aria-label="Switch to dark mode"  type="radio" name="__palette" id="__palette_0">
    
      <label class="md-header__button md-icon" title="Switch to dark mode" for="__palette_1" hidden>
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 7a5 5 0 0 1 5 5 5 5 0 0 1-5 5 5 5 0 0 1-5-5 5 5 0 0 1 5-5m0 2a3 3 0 0 0-3 3 3 3 0 0 0 3 3 3 3 0 0 0 3-3 3 3 0 0 0-3-3m0-7 2.39 3.42C13.65 5.15 12.84 5 12 5s-1.65.15-2.39.42zM3.34 7l4.16-.35A7.2 7.2 0 0 0 5.94 8.5c-.44.74-.69 1.5-.83 2.29zm.02 10 1.76-3.77a7.131 7.131 0 0 0 2.38 4.14zM20.65 7l-1.77 3.79a7.02 7.02 0 0 0-2.38-4.15zm-.01 10-4.14.36c.59-.51 1.12-1.14 1.54-1.86.42-.73.69-1.5.83-2.29zM12 22l-2.41-3.44c.74.27 1.55.44 2.41.44.82 0 1.63-.17 2.37-.44z"/></svg>
      </label>
    
  
    
    
    
    <input class="md-option" data-md-color-media="" data-md-color-scheme="slate" data-md-color-primary="indigo" data-md-color-accent="blue"  aria-label="Switch to light mode"  type="radio" name="__palette" id="__palette_1">
    
      <label class="md-header__button md-icon" title="Switch to light mode" for="__palette_0" hidden>
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="m17.75 4.09-2.53 1.94.91 3.06-2.63-1.81-2.63 1.81.91-3.06-2.53-1.94L12.44 4l1.06-3 1.06 3zm3.5 6.91-1.64 1.25.59 1.98-1.7-1.17-1.7 1.17.59-1.98L15.75 11l2.06-.05L18.5 9l.69 1.95zm-2.28 4.95c.83-.08 1.72 1.1 1.19 1.85-.32.45-.66.87-1.08 1.27C15.17 23 8.84 23 4.94 19.07c-3.91-3.9-3.91-10.24 0-14.14.4-.4.82-.76 1.27-1.08.75-.53 1.93.36 1.85 1.19-.27 2.86.69 5.83 2.89 8.02a9.96 9.96 0 0 0 8.02 2.89m-1.64 2.02a12.08 12.08 0 0 1-7.8-3.47c-2.17-2.19-3.33-5-3.49-7.82-2.81 3.14-2.7 7.96.31 10.98 3.02 3.01 7.84 3.12 10.98.31"/></svg>
      </label>
    
  
</form>
      
    
    
      <script>var palette=__md_get("__palette");if(palette&&palette.color){if("(prefers-color-scheme)"===palette.color.media){var media=matchMedia("(prefers-color-scheme: light)"),input=document.querySelector(media.matches?"[data-md-color-media='(prefers-color-scheme: light)']":"[data-md-color-media='(prefers-color-scheme: dark)']");palette.color.media=input.getAttribute("data-md-color-media"),palette.color.scheme=input.getAttribute("data-md-color-scheme"),palette.color.primary=input.getAttribute("data-md-color-primary"),palette.color.accent=input.getAttribute("data-md-color-accent")}for(var[key,value]of Object.entries(palette.color))document.body.setAttribute("data-md-color-"+key,value)}</script>
    
    
      <div class="md-header__option">
  <div class="md-select">
    
    <button class="md-header__button md-icon" aria-label="Select language">
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="m12.87 15.07-2.54-2.51.03-.03A17.5 17.5 0 0 0 14.07 6H17V4h-7V2H8v2H1v2h11.17C11.5 7.92 10.44 9.75 9 11.35 8.07 10.32 7.3 9.19 6.69 8h-2c.73 1.63 1.73 3.17 2.98 4.56l-5.09 5.02L4 19l5-5 3.11 3.11zM18.5 10h-2L12 22h2l1.12-3h4.75L21 22h2zm-2.62 7 1.62-4.33L19.12 17z"/></svg>
    </button>
    <div class="md-select__inner">
      <ul class="md-select__list">
        
          <li class="md-select__item">
            <a href="./" hreflang="en" class="md-select__link">
              🌚 English
            </a>
          </li>
        
          <li class="md-select__item">
            <a href="../../zh/tutorials/single_cell/" hreflang="zh" class="md-select__link">
              🇨🇳 简体中文
            </a>
          </li>
        
      </ul>
    </div>
  </div>
</div>
    
    
      
      
        <label class="md-header__button md-icon" for="__search">
          
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.52 6.52 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5"/></svg>
        </label>
        <div class="md-search" data-md-component="search" role="dialog">
  <label class="md-search__overlay" for="__search"></label>
  <div class="md-search__inner" role="search">
    <form class="md-search__form" name="search">
      <input type="text" class="md-search__input" name="query" aria-label="Search" placeholder="Search" autocapitalize="off" autocorrect="off" autocomplete="off" spellcheck="false" data-md-component="search-query" required>
      <label class="md-search__icon md-icon" for="__search">
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.52 6.52 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5"/></svg>
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11z"/></svg>
      </label>
      <nav class="md-search__options" aria-label="Search">
        
        <button type="reset" class="md-search__icon md-icon" title="Clear" aria-label="Clear" tabindex="-1">
          
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"/></svg>
        </button>
      </nav>
      
    </form>
    <div class="md-search__output">
      <div class="md-search__scrollwrap" tabindex="0" data-md-scrollfix>
        <div class="md-search-result" data-md-component="search-result">
          <div class="md-search-result__meta">
            Initializing search
          </div>
          <ol class="md-search-result__list" role="presentation"></ol>
        </div>
      </div>
    </div>
  </div>
</div>
      
    
    
      <div class="md-header__source">
        <a href="https://github.com/immunomind/immunarch" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 7.1.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2025 Fonticons, Inc.--><path d="M439.6 236.1 244 40.5c-5.4-5.5-12.8-8.5-20.4-8.5s-15 3-20.4 8.4L162.5 81l51.5 51.5c27.1-9.1 52.7 16.8 43.4 43.7l49.7 49.7c34.2-11.8 61.2 31 35.5 56.7-26.5 26.5-70.2-2.9-56-37.3L240.3 199v121.9c25.3 12.5 22.3 41.8 9.1 55-6.4 6.4-15.2 10.1-24.3 10.1s-17.8-3.6-24.3-10.1c-17.6-17.6-11.1-46.9 11.2-56v-123c-20.8-8.5-24.6-30.7-18.6-45L142.6 101 8.5 235.1C3 240.6 0 247.9 0 255.5s3 15 8.5 20.4l195.6 195.7c5.4 5.4 12.7 8.4 20.4 8.4s15-3 20.4-8.4l194.7-194.7c5.4-5.4 8.4-12.8 8.4-20.4s-3-15-8.4-20.4"/></svg>
  </div>
  <div class="md-source__repository">
    immunomind/immunarch
  </div>
</a>
      </div>
    
  </nav>
  
    
      
<nav class="md-tabs" aria-label="Tabs" data-md-component="tabs">
  <div class="md-grid">
    <ul class="md-tabs__list">
      
        
  
  
  
  
    
    
      <li class="md-tabs__item">
        <a href="../.." class="md-tabs__link">
          
  
  
  Getting started

        </a>
      </li>
    
  

      
        
  
  
  
  
    
    
      <li class="md-tabs__item">
        <a href="../../concepts/glossary/" class="md-tabs__link">
          
  
  
  Concepts

        </a>
      </li>
    
  

      
        
  
  
  
  
    
    
      
  
  
  
  
    
    
      <li class="md-tabs__item">
        <a href="../../guides/data_schema/" class="md-tabs__link">
          
  
  
  User guide

        </a>
      </li>
    
  

    
  

      
        
  
  
  
  
    
    
      
  
  
  
  
    
    
      <li class="md-tabs__item">
        <a href="../../howto/read_airr_bulk/" class="md-tabs__link">
          
  
  
  How-to

        </a>
      </li>
    
  

    
  

      
        
  
  
  
    
  
  
    
    
      <li class="md-tabs__item md-tabs__item--active">
        <a href="./" class="md-tabs__link">
          
  
  
  Tutorials

        </a>
      </li>
    
  

      
        
  
  
  
  
    
    
      <li class="md-tabs__item">
        <a href="../../api/reference/" class="md-tabs__link">
          
  
  
  API reference

        </a>
      </li>
    
  

      
    </ul>
  </div>
</nav>
    
  
</header>
    
    <div class="md-container" data-md-component="container">
      
      
        
      
      <main class="md-main" data-md-component="main">
        <div class="md-main__inner md-grid">
          
            
              
              <div class="md-sidebar md-sidebar--primary" data-md-component="sidebar" data-md-type="navigation" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    


  


<nav class="md-nav md-nav--primary md-nav--lifted" aria-label="Navigation" data-md-level="0">
  <label class="md-nav__title" for="__drawer">
    <a href="../.." title="immunomind." class="md-nav__button md-logo" aria-label="immunomind." data-md-component="logo">
      
  <img src="../../assets/phoenix_cl2.png" alt="logo">

    </a>
    immunomind.
  </label>
  
    <div class="md-nav__source">
      <a href="https://github.com/immunomind/immunarch" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 7.1.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2025 Fonticons, Inc.--><path d="M439.6 236.1 244 40.5c-5.4-5.5-12.8-8.5-20.4-8.5s-15 3-20.4 8.4L162.5 81l51.5 51.5c27.1-9.1 52.7 16.8 43.4 43.7l49.7 49.7c34.2-11.8 61.2 31 35.5 56.7-26.5 26.5-70.2-2.9-56-37.3L240.3 199v121.9c25.3 12.5 22.3 41.8 9.1 55-6.4 6.4-15.2 10.1-24.3 10.1s-17.8-3.6-24.3-10.1c-17.6-17.6-11.1-46.9 11.2-56v-123c-20.8-8.5-24.6-30.7-18.6-45L142.6 101 8.5 235.1C3 240.6 0 247.9 0 255.5s3 15 8.5 20.4l195.6 195.7c5.4 5.4 12.7 8.4 20.4 8.4s15-3 20.4-8.4l194.7-194.7c5.4-5.4 8.4-12.8 8.4-20.4s-3-15-8.4-20.4"/></svg>
  </div>
  <div class="md-source__repository">
    immunomind/immunarch
  </div>
</a>
    </div>
  
  <ul class="md-nav__list" data-md-scrollfix>
    
      
      
  
  
  
  
    
    
    
    
      
      
        
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
          
        
        <input class="md-nav__toggle md-toggle md-toggle--indeterminate" type="checkbox" id="__nav_1" >
        
          
          <label class="md-nav__link" for="__nav_1" id="__nav_1_label" tabindex="0">
            
  
  
  <span class="md-ellipsis">
    Getting started
    
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_1_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_1">
            <span class="md-nav__icon md-icon"></span>
            Getting started
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../.." class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Home
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../intro/installation/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Installation
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../intro/quick_start/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Quick Start
    
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
      
      
  
  
  
  
    
    
    
    
      
      
        
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
          
        
        <input class="md-nav__toggle md-toggle md-toggle--indeterminate" type="checkbox" id="__nav_2" >
        
          
          <label class="md-nav__link" for="__nav_2" id="__nav_2_label" tabindex="0">
            
  
  
  <span class="md-ellipsis">
    Concepts
    
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_2_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_2">
            <span class="md-nav__icon md-icon"></span>
            Concepts
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../concepts/glossary/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Glossary
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    
    
    
    
      
      
        
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
          
        
        <input class="md-nav__toggle md-toggle md-toggle--indeterminate" type="checkbox" id="__nav_2_2" >
        
          
          <label class="md-nav__link" for="__nav_2_2" id="__nav_2_2_label" tabindex="0">
            
  
  
  <span class="md-ellipsis">
    Basics
    
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="2" aria-labelledby="__nav_2_2_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_2_2">
            <span class="md-nav__icon md-icon"></span>
            Basics
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../concepts/basics/overview/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Overview
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../concepts/basics/data_units/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Units: chain -&gt; barcode -&gt; receptor
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../concepts/basics/aggregation/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Aggregation: defining receptors and repertoires
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../concepts/basics/immutability/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Pipeline-based execution: immutability and materialisation
    
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    
    
    
    
      
      
        
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
          
        
        <input class="md-nav__toggle md-toggle md-toggle--indeterminate" type="checkbox" id="__nav_2_3" >
        
          
          <label class="md-nav__link" for="__nav_2_3" id="__nav_2_3_label" tabindex="0">
            
  
  
  <span class="md-ellipsis">
    Workflow
    
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="2" aria-labelledby="__nav_2_3_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_2_3">
            <span class="md-nav__icon md-icon"></span>
            Workflow
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../concepts/workflow/phase_ingestion/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Phase 1: Ingestion
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../concepts/workflow/phase_transformation/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Phase 2: Transformation
    
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
      
      
  
  
  
  
    
    
    
    
      
      
        
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
          
        
        <input class="md-nav__toggle md-toggle md-toggle--indeterminate" type="checkbox" id="__nav_3" >
        
          
          <label class="md-nav__link" for="__nav_3" id="__nav_3_label" tabindex="0">
            
  
  
  <span class="md-ellipsis">
    User guide
    
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_3_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_3">
            <span class="md-nav__icon md-icon"></span>
            User guide
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    
    
    
    
      
      
        
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
          
        
        <input class="md-nav__toggle md-toggle md-toggle--indeterminate" type="checkbox" id="__nav_3_1" >
        
          
          <label class="md-nav__link" for="__nav_3_1" id="__nav_3_1_label" tabindex="0">
            
  
  
  <span class="md-ellipsis">
    Data structure
    
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="2" aria-labelledby="__nav_3_1_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_3_1">
            <span class="md-nav__icon md-icon"></span>
            Data structure
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../guides/data_schema/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    ImmunData structure
    
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../guides/faq/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    F.A.Q.
    
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
      
      
  
  
  
  
    
    
    
    
      
      
        
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
          
        
        <input class="md-nav__toggle md-toggle md-toggle--indeterminate" type="checkbox" id="__nav_4" >
        
          
          <label class="md-nav__link" for="__nav_4" id="__nav_4_label" tabindex="0">
            
  
  
  <span class="md-ellipsis">
    How-to
    
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_4_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_4">
            <span class="md-nav__icon md-icon"></span>
            How-to
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    
    
    
    
      
      
        
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
          
        
        <input class="md-nav__toggle md-toggle md-toggle--indeterminate" type="checkbox" id="__nav_4_1" >
        
          
          <label class="md-nav__link" for="__nav_4_1" id="__nav_4_1_label" tabindex="0">
            
  
  
  <span class="md-ellipsis">
    Input
    
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="2" aria-labelledby="__nav_4_1_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_4_1">
            <span class="md-nav__icon md-icon"></span>
            Input
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../howto/read_airr_bulk/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Bulk AIRR-C
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../howto/read_airr_sc/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Single-cell AIRR-C
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../howto/immunarch_import/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Import old `immunarch` data
    
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    
    
    
    
      
      
        
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
          
        
        <input class="md-nav__toggle md-toggle md-toggle--indeterminate" type="checkbox" id="__nav_4_2" >
        
          
          <label class="md-nav__link" for="__nav_4_2" id="__nav_4_2_label" tabindex="0">
            
  
  
  <span class="md-ellipsis">
    Performance
    
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="2" aria-labelledby="__nav_4_2_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_4_2">
            <span class="md-nav__icon md-icon"></span>
            Performance
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../howto/no_threads/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Limit the number of threads used by immunarch (DuckDB)
    
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
      
      
  
  
    
  
  
  
    
    
    
    
      
        
        
      
      
        
      
    
    
    <li class="md-nav__item md-nav__item--active md-nav__item--section md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_5" checked>
        
          
          <label class="md-nav__link" for="__nav_5" id="__nav_5_label" tabindex="">
            
  
  
  <span class="md-ellipsis">
    Tutorials
    
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_5_label" aria-expanded="true">
          <label class="md-nav__title" for="__nav_5">
            <span class="md-nav__icon md-icon"></span>
            Tutorials
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
    
  
  
  
    <li class="md-nav__item md-nav__item--active">
      
      <input class="md-nav__toggle md-toggle" type="checkbox" id="__toc">
      
      
      
        <label class="md-nav__link md-nav__link--active" for="__toc">
          
  
  
  <span class="md-ellipsis">
    Comprehensive analysis of single-cell immune repertoire data using `immunarch` and `immundata`
    
  </span>
  

          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <a href="./" class="md-nav__link md-nav__link--active">
        
  
  
  <span class="md-ellipsis">
    Comprehensive analysis of single-cell immune repertoire data using `immunarch` and `immundata`
    
  </span>
  

      </a>
      
        

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#introduction" class="md-nav__link">
    <span class="md-ellipsis">
      Introduction
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Introduction">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#what-youll-learn" class="md-nav__link">
    <span class="md-ellipsis">
      What you'll learn
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#dataset-used-in-this-tutorial" class="md-nav__link">
    <span class="md-ellipsis">
      Dataset used in this tutorial
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#helpful-links" class="md-nav__link">
    <span class="md-ellipsis">
      Helpful links
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#setup" class="md-nav__link">
    <span class="md-ellipsis">
      Setup
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Setup">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#pre-requisites" class="md-nav__link">
    <span class="md-ellipsis">
      Pre-requisites
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#using-pak" class="md-nav__link">
    <span class="md-ellipsis">
      Using pak
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#using-cran" class="md-nav__link">
    <span class="md-ellipsis">
      Using CRAN
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#load-necessary-packages-and-setup" class="md-nav__link">
    <span class="md-ellipsis">
      Load necessary packages and setup
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#typical-workflow-from-files-to-analysis" class="md-nav__link">
    <span class="md-ellipsis">
      Typical workflow: from files to analysis
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Typical workflow: from files to analysis">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#note-on-materialization" class="md-nav__link">
    <span class="md-ellipsis">
      Note on materialization
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#working-with-data" class="md-nav__link">
    <span class="md-ellipsis">
      Working with data
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Working with data">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#load-airr-data" class="md-nav__link">
    <span class="md-ellipsis">
      Load AIRR data
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#load-scrnaseq-data" class="md-nav__link">
    <span class="md-ellipsis">
      Load scRNAseq data
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#how-to-read-other-types-of-data-bulk-single-chain-etc" class="md-nav__link">
    <span class="md-ellipsis">
      How to read other types of data --- bulk, single-chain, etc.
    </span>
  </a>
  
    <nav class="md-nav" aria-label="How to read other types of data --- bulk, single-chain, etc.">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#1-bulk-sequencing" class="md-nav__link">
    <span class="md-ellipsis">
      1. Bulk sequencing
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#2-single-cell-sequencing" class="md-nav__link">
    <span class="md-ellipsis">
      2. Single-cell sequencing
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#3-airr-c-or-10x-genomics-format" class="md-nav__link">
    <span class="md-ellipsis">
      3. AIRR-C or 10x Genomics format
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#4-reading-file-paths-from-metadata" class="md-nav__link">
    <span class="md-ellipsis">
      4. Reading file paths from metadata
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#analyse-immune-repertoires" class="md-nav__link">
    <span class="md-ellipsis">
      Analyse immune repertoires
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Analyse immune repertoires">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#key-immune-repertoire-statistics" class="md-nav__link">
    <span class="md-ellipsis">
      Key immune repertoire statistics
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Key immune repertoire statistics">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#1a-number-of-receptors-and-barcodes-for-samples" class="md-nav__link">
    <span class="md-ellipsis">
      1a. Number of receptors and barcodes for samples
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#1b-number-of-receptors-and-barcodes-for-cell-clusters" class="md-nav__link">
    <span class="md-ellipsis">
      1b. Number of receptors and barcodes for cell clusters
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#2a-global-gene-usage" class="md-nav__link">
    <span class="md-ellipsis">
      2a. Global gene usage
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#2b-most-overrepresented-v-gene-segments" class="md-nav__link">
    <span class="md-ellipsis">
      2b. Most overrepresented V gene segments
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#2c-cluster-specific-v-gene-distribution" class="md-nav__link">
    <span class="md-ellipsis">
      2c. Cluster-specific V gene distribution
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#3a-cdr3-length-distribution-for-the-full-dataset" class="md-nav__link">
    <span class="md-ellipsis">
      3a. CDR3 length distribution for the full dataset
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#3b-cdr3-length-distribution-for-the-selected-clusters" class="md-nav__link">
    <span class="md-ellipsis">
      3b. CDR3 length distribution for the selected clusters
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
        
          <li class="md-nav__item">
  <a href="#clonality-quantifying-immune-expansion" class="md-nav__link">
    <span class="md-ellipsis">
      Clonality: quantifying immune expansion
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Clonality: quantifying immune expansion">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#1-rank-abundance" class="md-nav__link">
    <span class="md-ellipsis">
      1. Rank abundance
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#2-clonal-space-partitioning" class="md-nav__link">
    <span class="md-ellipsis">
      2. Clonal space partitioning
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#3-clonal-rank-bins" class="md-nav__link">
    <span class="md-ellipsis">
      3. Clonal rank bins
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
        
          <li class="md-nav__item">
  <a href="#diversity-how-rich-and-even-are-your-repertoires" class="md-nav__link">
    <span class="md-ellipsis">
      Diversity: how rich and even are your repertoires?
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#public-receptor-indices-quantifying-repertoire-overlap" class="md-nav__link">
    <span class="md-ellipsis">
      Public receptor indices: quantifying repertoire overlap
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#discover-and-annotate-immune-receptors" class="md-nav__link">
    <span class="md-ellipsis">
      Discover and annotate immune receptors
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Discover and annotate immune receptors">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#annotating-receptors-by-clonality" class="md-nav__link">
    <span class="md-ellipsis">
      Annotating receptors by clonality
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#other-sections" class="md-nav__link">
    <span class="md-ellipsis">
      Other sections
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#wip-analyse-subgroups-of-receptors-of-interest" class="md-nav__link">
    <span class="md-ellipsis">
      [WIP] Analyse subgroups of receptors of interest
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#wip-conclusion" class="md-nav__link">
    <span class="md-ellipsis">
      [WIP] Conclusion
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#testing-zone" class="md-nav__link">
    <span class="md-ellipsis">
      Testing zone
    </span>
  </a>
  
</li>
      
    </ul>
  
</nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../migration/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Migration from immunarch v0.9 to v1.0
    
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
      
      
  
  
  
  
    
    
    
    
      
      
        
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
          
        
        <input class="md-nav__toggle md-toggle md-toggle--indeterminate" type="checkbox" id="__nav_6" >
        
          
          <label class="md-nav__link" for="__nav_6" id="__nav_6_label" tabindex="0">
            
  
  
  <span class="md-ellipsis">
    API reference
    
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_6_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_6">
            <span class="md-nav__icon md-icon"></span>
            API reference
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../api/reference/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    API reference
    
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
  </ul>
</nav>
                  </div>
                </div>
              </div>
            
            
              
              <div class="md-sidebar md-sidebar--secondary" data-md-component="sidebar" data-md-type="toc" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#introduction" class="md-nav__link">
    <span class="md-ellipsis">
      Introduction
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Introduction">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#what-youll-learn" class="md-nav__link">
    <span class="md-ellipsis">
      What you'll learn
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#dataset-used-in-this-tutorial" class="md-nav__link">
    <span class="md-ellipsis">
      Dataset used in this tutorial
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#helpful-links" class="md-nav__link">
    <span class="md-ellipsis">
      Helpful links
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#setup" class="md-nav__link">
    <span class="md-ellipsis">
      Setup
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Setup">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#pre-requisites" class="md-nav__link">
    <span class="md-ellipsis">
      Pre-requisites
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#using-pak" class="md-nav__link">
    <span class="md-ellipsis">
      Using pak
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#using-cran" class="md-nav__link">
    <span class="md-ellipsis">
      Using CRAN
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#load-necessary-packages-and-setup" class="md-nav__link">
    <span class="md-ellipsis">
      Load necessary packages and setup
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#typical-workflow-from-files-to-analysis" class="md-nav__link">
    <span class="md-ellipsis">
      Typical workflow: from files to analysis
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Typical workflow: from files to analysis">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#note-on-materialization" class="md-nav__link">
    <span class="md-ellipsis">
      Note on materialization
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#working-with-data" class="md-nav__link">
    <span class="md-ellipsis">
      Working with data
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Working with data">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#load-airr-data" class="md-nav__link">
    <span class="md-ellipsis">
      Load AIRR data
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#load-scrnaseq-data" class="md-nav__link">
    <span class="md-ellipsis">
      Load scRNAseq data
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#how-to-read-other-types-of-data-bulk-single-chain-etc" class="md-nav__link">
    <span class="md-ellipsis">
      How to read other types of data --- bulk, single-chain, etc.
    </span>
  </a>
  
    <nav class="md-nav" aria-label="How to read other types of data --- bulk, single-chain, etc.">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#1-bulk-sequencing" class="md-nav__link">
    <span class="md-ellipsis">
      1. Bulk sequencing
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#2-single-cell-sequencing" class="md-nav__link">
    <span class="md-ellipsis">
      2. Single-cell sequencing
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#3-airr-c-or-10x-genomics-format" class="md-nav__link">
    <span class="md-ellipsis">
      3. AIRR-C or 10x Genomics format
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#4-reading-file-paths-from-metadata" class="md-nav__link">
    <span class="md-ellipsis">
      4. Reading file paths from metadata
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#analyse-immune-repertoires" class="md-nav__link">
    <span class="md-ellipsis">
      Analyse immune repertoires
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Analyse immune repertoires">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#key-immune-repertoire-statistics" class="md-nav__link">
    <span class="md-ellipsis">
      Key immune repertoire statistics
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Key immune repertoire statistics">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#1a-number-of-receptors-and-barcodes-for-samples" class="md-nav__link">
    <span class="md-ellipsis">
      1a. Number of receptors and barcodes for samples
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#1b-number-of-receptors-and-barcodes-for-cell-clusters" class="md-nav__link">
    <span class="md-ellipsis">
      1b. Number of receptors and barcodes for cell clusters
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#2a-global-gene-usage" class="md-nav__link">
    <span class="md-ellipsis">
      2a. Global gene usage
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#2b-most-overrepresented-v-gene-segments" class="md-nav__link">
    <span class="md-ellipsis">
      2b. Most overrepresented V gene segments
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#2c-cluster-specific-v-gene-distribution" class="md-nav__link">
    <span class="md-ellipsis">
      2c. Cluster-specific V gene distribution
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#3a-cdr3-length-distribution-for-the-full-dataset" class="md-nav__link">
    <span class="md-ellipsis">
      3a. CDR3 length distribution for the full dataset
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#3b-cdr3-length-distribution-for-the-selected-clusters" class="md-nav__link">
    <span class="md-ellipsis">
      3b. CDR3 length distribution for the selected clusters
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
        
          <li class="md-nav__item">
  <a href="#clonality-quantifying-immune-expansion" class="md-nav__link">
    <span class="md-ellipsis">
      Clonality: quantifying immune expansion
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Clonality: quantifying immune expansion">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#1-rank-abundance" class="md-nav__link">
    <span class="md-ellipsis">
      1. Rank abundance
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#2-clonal-space-partitioning" class="md-nav__link">
    <span class="md-ellipsis">
      2. Clonal space partitioning
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#3-clonal-rank-bins" class="md-nav__link">
    <span class="md-ellipsis">
      3. Clonal rank bins
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
        
          <li class="md-nav__item">
  <a href="#diversity-how-rich-and-even-are-your-repertoires" class="md-nav__link">
    <span class="md-ellipsis">
      Diversity: how rich and even are your repertoires?
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#public-receptor-indices-quantifying-repertoire-overlap" class="md-nav__link">
    <span class="md-ellipsis">
      Public receptor indices: quantifying repertoire overlap
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#discover-and-annotate-immune-receptors" class="md-nav__link">
    <span class="md-ellipsis">
      Discover and annotate immune receptors
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Discover and annotate immune receptors">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#annotating-receptors-by-clonality" class="md-nav__link">
    <span class="md-ellipsis">
      Annotating receptors by clonality
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#other-sections" class="md-nav__link">
    <span class="md-ellipsis">
      Other sections
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#wip-analyse-subgroups-of-receptors-of-interest" class="md-nav__link">
    <span class="md-ellipsis">
      [WIP] Analyse subgroups of receptors of interest
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#wip-conclusion" class="md-nav__link">
    <span class="md-ellipsis">
      [WIP] Conclusion
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#testing-zone" class="md-nav__link">
    <span class="md-ellipsis">
      Testing zone
    </span>
  </a>
  
</li>
      
    </ul>
  
</nav>
                  </div>
                </div>
              </div>
            
          
          
            <div class="md-content" data-md-component="content">
              <article class="md-content__inner md-typeset">
                
                  


  
  


  <h1>Comprehensive analysis of single-cell immune repertoire data using `immunarch` and `immundata`</h1>

<h2 id="introduction">Introduction</h2>
<p>This vignette is a practical, end‑to‑end guide to analyzing single‑chain
and paired‑chain Adaptive Immune Receptor Repertoire (AIRR) data derived
from single‑cell assays (scRNA‑seq with VDJ; also referred to as
scVDJ‑seq / scTCR‑seq / scBCR‑seq). It starts from the basics and builds
up to more advanced analyses, so it can serve both as a first stop for
newcomers and as a focused introduction for experienced readers who want
to learn <strong><code>immunarch</code></strong> and <strong><code>immundata</code></strong> specifically.</p>
<h3 id="what-youll-learn">What you'll learn</h3>
<ul>
<li>Load single‑cell V(D)J data and metadata into
    <code>immunarch</code>/<code>immundata</code>.</li>
<li>Define clonotypes (AA vs NT), set chain‑pairing rules, and perform
    essential QC.</li>
<li>Explore repertoire features: expansion, gene usage, CDR3 properties,
    and diversity.</li>
<li>Quantify overlap/similarity between samples and conditions.</li>
<li>Link clonotypes to single‑cell transcriptomic clusters for
    downstream interpretation.</li>
</ul>
<h3 id="dataset-used-in-this-tutorial">Dataset used in this tutorial</h3>
<p>We will work with data for <strong>patient 6</strong> from the study <strong>"Peripheral T
cell expansion predicts tumour infiltration and clinical response"</strong> (Wu
<em>et al.</em>; <em>Nature</em> 579:274--278, 2020). The dataset includes three
compartments per patient: <em>tumour</em>, <em>peripheral blood</em>, and <em>normal
adjacent tissue (NAT)</em>. We will demonstrate a typical single‑cell AIRR
workflow across these three samples. Bulletpoint description of this
great study:</p>
<ul>
<li>Deep single‑cell RNA/TCR profiling across tumour, NAT, and blood</li>
<li>Clonotypic expansion of effector‑like T cells in and around tumours</li>
<li>Ppatients with stronger peripheral expansion signatures tend to
    respond better to anti‑PD‑(L)1 therapy</li>
<li>Expanded intratumoural clonotypes are often detectable in blood,
    enabling convenient monitoring</li>
</ul>
<p><strong>References &amp; data access:</strong></p>
<ul>
<li>Wu TD, Madireddi S, de Almeida PE, <em>et al.</em> <em>Nature</em> (2020).
    "Peripheral T cell expansion predicts tumour infiltration and
    clinical response."</li>
<li>GEO series: <strong>GSE139555</strong> (single‑cell RNA/TCR; pretreatment samples
    from 14 patients; tumour/NAT/blood).</li>
</ul>
<h3 id="helpful-links">Helpful links</h3>
<ul>
<li>Source code and issue tracker:
    <a href="https://github.com/immunomind/immunarch"><code>immunarch</code></a> ·
    <a href="https://github.com/immunomind/immundata"><code>immundata</code></a></li>
<li>Interactive assistant: <strong>AIRR‑GPT</strong> --- a GPT Store model preloaded
    with <code>immunarch</code> and <code>immundata</code> docs to help you as you go:
    <a href="https://chatgpt.com/g/g-67fe54e302348191814e4ef2efb3bad5-airr-gpt">https://chatgpt.com/g/g-67fe54e302348191814e4ef2efb3bad5-airr-gpt</a></li>
</ul>
<hr />
<h2 id="setup">Setup</h2>
<div class="tabbed-set tabbed-alternate" data-tabs="1:1"><input checked="checked" id="__tabbed_1_1" name="__tabbed_1" type="radio" /><div class="tabbed-labels"><label for="__tabbed_1_1">R</label></div>
<div class="tabbed-content">
<div class="tabbed-block">
<h3 id="pre-requisites">Pre-requisites</h3>
<ul>
<li><strong>R</strong> ≥ 4.2 (4.3+ recommended)</li>
<li>Internet access to install packages and download data from GEO</li>
<li>\~2–4 GB RAM free for the small demonstration subset used here</li>
</ul>
<p>You can install the required packages with <strong>pak</strong> (recommended for reproducibility and fast binary installs) or with base installers.</p>
</div>
</div>
</div>
<h3 id="using-pak">Using <code>pak</code></h3>
<div class="tabbed-set tabbed-alternate" data-tabs="2:1"><input checked="checked" id="__tabbed_2_1" name="__tabbed_2" type="radio" /><div class="tabbed-labels"><label for="__tabbed_2_1">R</label></div>
<div class="tabbed-content">
<div class="tabbed-block">
<div class="cell-code highlight"><pre><span></span><code><span class="nf">install.packages</span><span class="p">(</span><span class="s">&quot;pak&quot;</span><span class="p">,</span><span class="w"> </span><span class="n">repos</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">sprintf</span><span class="p">(</span><span class="s">&quot;https://r-lib.github.io/p/pak/stable/%s/%s/%s&quot;</span><span class="p">,</span><span class="w"> </span><span class="n">.Platform</span><span class="o">$</span><span class="n">pkgType</span><span class="p">,</span><span class="w"> </span><span class="nf">R.Version</span><span class="p">()</span><span class="o">$</span><span class="n">os</span><span class="p">,</span><span class="w"> </span><span class="nf">R.Version</span><span class="p">()</span><span class="o">$</span><span class="n">arch</span><span class="p">))</span>

<span class="n">pak</span><span class="o">::</span><span class="nf">pkg_install</span><span class="p">(</span><span class="s">&quot;immunomind/immunarch&quot;</span><span class="p">)</span>
<span class="n">pak</span><span class="o">::</span><span class="nf">pkg_install</span><span class="p">(</span><span class="s">&quot;Seurat&quot;</span><span class="p">)</span>
<span class="n">pak</span><span class="o">::</span><span class="nf">pkg_install</span><span class="p">(</span><span class="s">&quot;BiocFileCache&quot;</span><span class="p">)</span>
<span class="n">pak</span><span class="o">::</span><span class="nf">pkg_install</span><span class="p">(</span><span class="s">&quot;ggthemes&quot;</span><span class="p">)</span>
<span class="n">pak</span><span class="o">::</span><span class="nf">pkg_install</span><span class="p">(</span><span class="s">&quot;ggsci&quot;</span><span class="p">)</span>
</code></pre></div>
</div>
</div>
</div>
<h3 id="using-cran">Using CRAN</h3>
<div class="tabbed-set tabbed-alternate" data-tabs="3:1"><input checked="checked" id="__tabbed_3_1" name="__tabbed_3" type="radio" /><div class="tabbed-labels"><label for="__tabbed_3_1">R</label></div>
<div class="tabbed-content">
<div class="tabbed-block">
<div class="cell-code highlight"><pre><span></span><code><span class="nf">install.packages</span><span class="p">(</span><span class="nf">c</span><span class="p">(</span><span class="s">&quot;immunarch&quot;</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;Seurat&quot;</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;BiocFileCache&quot;</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;ggthemes&quot;</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;ggsci&quot;</span><span class="p">))</span>
</code></pre></div>
</div>
</div>
</div>
<blockquote>
<p><em>Tip:</em> If you prefer fully reproducible environments, consider using
<code>renv::init()</code> to snapshot package versions.</p>
</blockquote>
<h3 id="load-necessary-packages-and-setup">Load necessary packages and setup</h3>
<div class="tabbed-set tabbed-alternate" data-tabs="4:1"><input checked="checked" id="__tabbed_4_1" name="__tabbed_4" type="radio" /><div class="tabbed-labels"><label for="__tabbed_4_1">R</label></div>
<div class="tabbed-content">
<div class="tabbed-block">
<div class="cell-code highlight"><pre><span></span><code><span class="c1"># Silence noisy startup messages while loading</span>
<span class="nf">suppressPackageStartupMessages</span><span class="p">({</span>
<span class="w">  </span><span class="nf">library</span><span class="p">(</span><span class="n">immunarch</span><span class="p">)</span>
<span class="w">  </span><span class="nf">library</span><span class="p">(</span><span class="n">Seurat</span><span class="p">)</span>
<span class="w">  </span><span class="nf">library</span><span class="p">(</span><span class="n">BiocFileCache</span><span class="p">)</span>
<span class="w">  </span><span class="nf">library</span><span class="p">(</span><span class="n">ggplot2</span><span class="p">)</span>
<span class="w">  </span><span class="nf">library</span><span class="p">(</span><span class="n">ggthemes</span><span class="p">)</span>
<span class="w">  </span><span class="nf">library</span><span class="p">(</span><span class="n">ggsci</span><span class="p">)</span>
<span class="p">})</span>

<span class="nf">theme_set</span><span class="p">(</span><span class="n">ggthemes</span><span class="o">::</span><span class="nf">theme_few</span><span class="p">())</span>

<span class="c1"># Keep a snapshot of versions for reproducibility (printed at the end as well)</span>
<span class="n">pkg_versions</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">list</span><span class="p">(</span>
<span class="w">  </span><span class="n">immunarch</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">as.character</span><span class="p">(</span><span class="n">utils</span><span class="o">::</span><span class="nf">packageVersion</span><span class="p">(</span><span class="s">&quot;immunarch&quot;</span><span class="p">)),</span>
<span class="w">  </span><span class="n">immundata</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">as.character</span><span class="p">(</span><span class="n">utils</span><span class="o">::</span><span class="nf">packageVersion</span><span class="p">(</span><span class="s">&quot;immundata&quot;</span><span class="p">)),</span>
<span class="w">  </span><span class="n">Seurat</span><span class="w">    </span><span class="o">=</span><span class="w"> </span><span class="nf">as.character</span><span class="p">(</span><span class="n">utils</span><span class="o">::</span><span class="nf">packageVersion</span><span class="p">(</span><span class="s">&quot;Seurat&quot;</span><span class="p">))</span>
<span class="p">)</span>

<span class="n">pkg_versions</span>
</code></pre></div>
<div class="highlight"><pre><span></span><code>$immunarch
[1] &quot;0.10.3&quot;

$immundata
[1] &quot;0.0.5&quot;

$Seurat
[1] &quot;5.3.0&quot;
</code></pre></div>
</div>
</div>
</div>
<hr />
<h2 id="typical-workflow-from-files-to-analysis">Typical workflow: from files to analysis</h2>
<p>Analyzing single-cell AIRR data with <code>immunarch 1.0</code> powered by
<code>immundata</code> follows a clear, reproducible workflow that separates data
ingestion from downstream transformation and annotation. This design
enables you to process and reuse large datasets efficiently without
manual reloading or reprocessing.</p>
<p>The general workflow is:</p>
<ol>
<li>
<p><strong>Ingestion phase</strong></p>
<ul>
<li>Read metadata (optional but recommended)</li>
<li>Load AIRR-seq files and preprocess as needed</li>
<li>Define receptors using a schema (single or paired chains, custom
    features)</li>
<li>Aggregate and persist the dataset as an immutable on-disk
    <code>ImmunData</code> object</li>
</ul>
</li>
<li>
<p><strong>Transformation phase</strong></p>
<ul>
<li>Annotate <code>ImmunData</code> with external information (e.g., scRNA-seq
    clusters)</li>
<li>Regroup repertoires on the fly (e.g., by tissue, donor, or cell
    state)</li>
<li>Filter, mutate, and compute statistics without copying or
    reloading data</li>
<li>Visualize results, save plots, or export annotations back to
    Seurat/AnnData</li>
</ul>
</li>
</ol>
<p>This separation of phases lets you:</p>
<ul>
<li>Ingest and persist raw AIRR-seq data once, then reload instantly for
    any analysis</li>
<li>Dynamically regroup receptors or repertoires, and annotate data
    without touching the original files</li>
<li>Build fully reproducible pipelines, as each step leaves previous
    data untouched</li>
</ul>
<p>You can read more about the phases in
<a href="https://immunomind.github.io/docs/concepts/workflow/phase_ingestion/">Concepts</a>.</p>
<h3 id="note-on-materialization">Note on materialization</h3>
<p>The <code>ImmunData</code> object in <code>immundata</code> is designed to handle large,
out-of-memory datasets efficiently. Instead of loading everything into
RAM, it tracks computations and only runs them when the results are
actually needed---such as for plotting or summarizing data. This is
called lazy evaluation.</p>
<p>Materialization means executing all pending computations and bringing
the results into memory as a concrete table or data frame. This approach
is powerful for working with large datasets, but can be unfamiliar if
you're used to working entirely in-memory.</p>
<p>If you see an error like:</p>
<pre><code>Error: Materialization is disabled, use collect() or as_tibble() to materialize.
</code></pre>
<p>...it means your data hasn't been materialized yet (i.e., the
computation hasn't run and the results aren't present in R memory).</p>
<ul>
<li>
<p>For small results (e.g., summary stats, small tables), simply call
    <code>collect()</code> or <code>as_tibble()</code> at the end of your pipeline to
    materialize the data:</p>
<div class="highlight"><pre><span></span><code><span class="n">result</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">immunarch_function</span><span class="p">(</span><span class="n">idata</span><span class="p">)</span><span class="w"> </span><span class="o">|&gt;</span><span class="w"> </span><span class="nf">collect</span><span class="p">()</span>
</code></pre></div>
</li>
<li>
<p>For large datasets (e.g., a fully filtered <code>ImmunData</code> object),
    avoid materializing unless you're sure you need all data in
    memory---this could use significant RAM and slow down your workflow.</p>
</li>
</ul>
<blockquote>
<p><strong>Tip:</strong> Use materialization for final summaries, plots, or tables,
but keep your core ImmunData object on disk and in its lazy, efficient
format for most analyses.</p>
</blockquote>
<p>You can read more about the materialization in
<a href="https://immunomind.github.io/docs/concepts/immutability/">Concepts</a>.</p>
<hr />
<h2 id="working-with-data">Working with data</h2>
<h3 id="load-airr-data">Load AIRR data</h3>
<p>First, we need to load sample metadata. The metadata file is a
tab-separated table, where each row represents a sample (or file), and
columns contain sample-level features such as <code>"Tissue"</code>, <code>"Donor"</code>, or
<code>"ImmunotherapyResponse"</code>.</p>
<div class="tabbed-set tabbed-alternate" data-tabs="5:1"><input checked="checked" id="__tabbed_5_1" name="__tabbed_5" type="radio" /><div class="tabbed-labels"><label for="__tabbed_5_1">R</label></div>
<div class="tabbed-content">
<div class="tabbed-block">
<div class="cell-code highlight"><pre><span></span><code><span class="n">md_file</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">system.file</span><span class="p">(</span><span class="s">&quot;extdata/single_cell&quot;</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;metadata.tsv&quot;</span><span class="p">,</span><span class="w"> </span><span class="n">package</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;immundata&quot;</span><span class="p">)</span>
<span class="n">md_file</span>
</code></pre></div>
<div class="highlight"><pre><span></span><code>[1] &quot;/home/runner/work/_temp/Library/immundata/extdata/single_cell/metadata.tsv&quot;
</code></pre></div>
<div class="cell-code highlight"><pre><span></span><code><span class="n">md_table</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">read_metadata</span><span class="p">(</span><span class="n">md_file</span><span class="p">)</span>
</code></pre></div>
<div class="highlight"><pre><span></span><code>Rows: 3 Columns: 3
── Column specification ────────────────────────────────────────────────────────
Delimiter: &quot;\t&quot;
chr (3): File, Tissue, Prefix

ℹ Use `spec()` to retrieve the full column specification for this data.
ℹ Specify the column types or set `show_col_types = FALSE` to quiet this message.
ℹ Found 3/3 repertoire files from the metadata on the disk

✔ Metadata parsed successfully
</code></pre></div>
<div class="cell-code highlight"><pre><span></span><code><span class="n">md_table</span>
</code></pre></div>
<div class="highlight"><pre><span></span><code># A tibble: 3 × 4
  File                                                    Tissue Prefix filename
  &lt;chr&gt;                                                   &lt;chr&gt;  &lt;chr&gt;  &lt;chr&gt;   
1 /home/runner/work/_temp/Library/immundata/extdata/sing… Blood  LB6_   /home/r…
2 /home/runner/work/_temp/Library/immundata/extdata/sing… Normal LN6_   /home/r…
3 /home/runner/work/_temp/Library/immundata/extdata/sing… Tumor  LT6_   /home/r…
</code></pre></div>
</div>
</div>
</div>
<p>The <code>"Tissue"</code> column defines which biological compartment (tumour,
blood, or NAT) each sample comes from. We'll use it to split data into
repertoires. Later in this tutorial, you'll see how to dynamically split
data by other annotations, such as cell clusters from scRNA-seq.</p>
<p>Next, get the file paths to the V(D)J data packaged with <code>immundata</code>:</p>
<div class="tabbed-set tabbed-alternate" data-tabs="6:1"><input checked="checked" id="__tabbed_6_1" name="__tabbed_6" type="radio" /><div class="tabbed-labels"><label for="__tabbed_6_1">R</label></div>
<div class="tabbed-content">
<div class="tabbed-block">
<div class="cell-code highlight"><pre><span></span><code><span class="n">inp_files</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">paste0</span><span class="p">(</span><span class="nf">system.file</span><span class="p">(</span><span class="s">&quot;extdata/single_cell&quot;</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;&quot;</span><span class="p">,</span><span class="w"> </span><span class="n">package</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;immundata&quot;</span><span class="p">),</span><span class="w"> </span><span class="s">&quot;/*.csv.gz&quot;</span><span class="p">)</span>
<span class="n">inp_files</span>
</code></pre></div>
<div class="highlight"><pre><span></span><code>[1] &quot;/home/runner/work/_temp/Library/immundata/extdata/single_cell//*.csv.gz&quot;
</code></pre></div>
</div>
</div>
</div>
<p>What is a receptor in <code>immundata</code>? A receptor is the central unit of
data analysis in <code>immundata</code>. It is defined by:</p>
<ul>
<li>The features used for grouping (e.g., <code>cdr3</code>, <code>v_call</code>, <code>j_call</code>)</li>
<li>The chains involved (e.g., <code>TRB</code> for TCR beta, or paired
    <code>TRA</code>+<code>TRB</code>)</li>
</ul>
<p>Defining a receptor schema lets you flexibly group TCR/BCR sequences for
downstream analysis and compute statistics on them.</p>
<p>Here's how to define a basic single-chain receptor schema: TRB chains,
grouped by both <code>cdr3</code> and <code>v_call</code> (amino acid sequence and V gene).</p>
<div class="tabbed-set tabbed-alternate" data-tabs="7:1"><input checked="checked" id="__tabbed_7_1" name="__tabbed_7" type="radio" /><div class="tabbed-labels"><label for="__tabbed_7_1">R</label></div>
<div class="tabbed-content">
<div class="tabbed-block">
<div class="cell-code highlight"><pre><span></span><code><span class="n">schema</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">make_receptor_schema</span><span class="p">(</span><span class="n">features</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="s">&quot;cdr3&quot;</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;v_call&quot;</span><span class="p">),</span><span class="w"> </span><span class="n">chains</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="s">&quot;TRB&quot;</span><span class="p">))</span>
<span class="n">schema</span>
</code></pre></div>
<div class="highlight"><pre><span></span><code>$features
[1] &quot;cdr3&quot;   &quot;v_call&quot;

$chains
[1] &quot;TRB&quot;
</code></pre></div>
</div>
</div>
</div>
<p>If you want to use other receptor definitions, you can create and pass
custom schemas to <code>read_repertoires()</code>. Below are a few examples:</p>
<div class="tabbed-set tabbed-alternate" data-tabs="8:1"><input checked="checked" id="__tabbed_8_1" name="__tabbed_8" type="radio" /><div class="tabbed-labels"><label for="__tabbed_8_1">R</label></div>
<div class="tabbed-content">
<div class="tabbed-block">
<div class="cell-code highlight"><pre><span></span><code><span class="c1"># Paired-chain</span>
<span class="n">schema</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">make_receptor_schema</span><span class="p">(</span><span class="n">features</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="s">&quot;cdr3&quot;</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;v_call&quot;</span><span class="p">),</span><span class="w"> </span><span class="n">chains</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="s">&quot;TRA&quot;</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;TRB&quot;</span><span class="p">))</span>
<span class="n">schema</span>
</code></pre></div>
<div class="highlight"><pre><span></span><code>$features
[1] &quot;cdr3&quot;   &quot;v_call&quot;

$chains
[1] &quot;TRA&quot; &quot;TRB&quot;
</code></pre></div>
<div class="cell-code highlight"><pre><span></span><code><span class="c1"># Single-chain with &quot;cdr3&quot; only</span>
<span class="n">schema</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">make_receptor_schema</span><span class="p">(</span><span class="n">features</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="s">&quot;cdr3&quot;</span><span class="p">),</span><span class="w"> </span><span class="n">chains</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="s">&quot;TRA&quot;</span><span class="p">))</span>
<span class="n">schema</span>
</code></pre></div>
<div class="highlight"><pre><span></span><code>$features
[1] &quot;cdr3&quot;

$chains
[1] &quot;TRA&quot;
</code></pre></div>
<div class="cell-code highlight"><pre><span></span><code><span class="c1"># The most strict receptor definition</span>
<span class="n">schema</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">make_receptor_schema</span><span class="p">(</span><span class="n">features</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="s">&quot;cdr3&quot;</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;v_call&quot;</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;j_call&quot;</span><span class="p">),</span><span class="w"> </span><span class="n">chains</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="s">&quot;TRA&quot;</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;TRB&quot;</span><span class="p">))</span>
<span class="n">schema</span>
</code></pre></div>
<div class="highlight"><pre><span></span><code>$features
[1] &quot;cdr3&quot;   &quot;v_call&quot; &quot;j_call&quot;

$chains
[1] &quot;TRA&quot; &quot;TRB&quot;
</code></pre></div>
</div>
</div>
</div>
<p>The main entry point for loading AIRR-seq data into the <code>immunarch 1.0</code>
framework is the <code>read_repertoires()</code> function from <code>immundata</code>. This
function handles everything from reading raw files, to preprocessing and
aggregation, to joining metadata and saving an efficient on-disk
ImmunData object for reproducible analysis.</p>
<p>Below, we describe the key parameters you'll want to understand:</p>
<ul>
<li><strong><code>path</code></strong> --- vector of input file paths, e.g., to AIRR TSV, 10X
    CSV, or Parquet files. (You can use <code>Sys.glob()</code> to collect files.)</li>
<li><strong><code>schema</code></strong> --- defines how receptors are grouped; typically
    created with <code>make_receptor_schema()</code>.</li>
<li><strong><code>metadata</code></strong> --- optional data frame with sample-level metadata,
    read by <code>read_metadata()</code>.</li>
<li><strong><code>barcode_col</code></strong> --- name of the column containing cell barcodes
    (e.g., <code>"barcode"</code> for 10x single-cell data); this triggers
    single-cell logic.</li>
<li><strong><code>locus_col</code></strong> --- name of the column specifying the chain (e.g.,
    <code>"locus"</code> for TRA/TRB distinction).</li>
<li><strong><code>umi_col</code></strong> --- column for UMI counts (e.g., <code>"umis"</code>); used to
    select dominant chains per barcode.</li>
<li><strong><code>preprocess</code></strong> --- list of preprocessing steps to apply before
    aggregation. The preset <code>make_default_preprocessing("10x")</code> works
    for standard 10x data.</li>
<li><strong><code>repertoire_schema</code></strong> --- columns in metadata (or annotation) used
    to define repertoires (e.g., <code>"Tissue"</code> to split by compartment).</li>
</ul>
<p>Here's how to use it on the packaged single-cell demo dataset:</p>
<div class="tabbed-set tabbed-alternate" data-tabs="9:1"><input checked="checked" id="__tabbed_9_1" name="__tabbed_9" type="radio" /><div class="tabbed-labels"><label for="__tabbed_9_1">R</label></div>
<div class="tabbed-content">
<div class="tabbed-block">
<div class="cell-code highlight"><pre><span></span><code><span class="n">schema</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">make_receptor_schema</span><span class="p">(</span><span class="n">features</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="s">&quot;cdr3&quot;</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;v_call&quot;</span><span class="p">),</span><span class="w"> </span><span class="n">chains</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="s">&quot;TRB&quot;</span><span class="p">))</span>

<span class="n">idata</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">read_repertoires</span><span class="p">(</span><span class="n">path</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">inp_files</span><span class="p">,</span><span class="w"> </span>
<span class="w">                          </span><span class="n">schema</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">schema</span><span class="p">,</span><span class="w"> </span>
<span class="w">                          </span><span class="n">metadata</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">md_table</span><span class="p">,</span><span class="w"> </span>
<span class="w">                          </span><span class="n">barcode_col</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;barcode&quot;</span><span class="p">,</span><span class="w"> </span>
<span class="w">                          </span><span class="n">locus_col</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;locus&quot;</span><span class="p">,</span><span class="w">  </span><span class="c1"># Mind that we use `locus` here because of column renaming</span>
<span class="w">                          </span><span class="n">umi_col</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;umis&quot;</span><span class="p">,</span>
<span class="w">                          </span><span class="n">preprocess</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">make_default_preprocessing</span><span class="p">(</span><span class="s">&quot;10x&quot;</span><span class="p">),</span><span class="w"> </span>
<span class="w">                          </span><span class="n">repertoire_schema</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;Tissue&quot;</span><span class="p">)</span>
</code></pre></div>
<div class="highlight"><pre><span></span><code>
</code></pre></div>
<div class="highlight"><pre><span></span><code>── Reading repertoire data 
</code></pre></div>
<div class="highlight"><pre><span></span><code>  1. /home/runner/work/_temp/Library/immundata/extdata/single_cell/lb6.csv.gz
</code></pre></div>
<div class="highlight"><pre><span></span><code>  2. /home/runner/work/_temp/Library/immundata/extdata/single_cell/ln6.csv.gz
</code></pre></div>
<div class="highlight"><pre><span></span><code>  3. /home/runner/work/_temp/Library/immundata/extdata/single_cell/lt6.csv.gz
</code></pre></div>
<div class="highlight"><pre><span></span><code>ℹ Checking if all files are of the same type
</code></pre></div>
<div class="highlight"><pre><span></span><code>✔ All files have the same extension
</code></pre></div>
<div class="highlight"><pre><span></span><code>
</code></pre></div>
<div class="highlight"><pre><span></span><code>── Renaming the columns and schemas 
</code></pre></div>
<div class="highlight"><pre><span></span><code>✔ Introduced new renamed columns: locus, v_call, d_call, and j_call
</code></pre></div>
<div class="highlight"><pre><span></span><code>✔ Renaming is finished
</code></pre></div>
<div class="highlight"><pre><span></span><code>
</code></pre></div>
<div class="highlight"><pre><span></span><code>── Preprocessing the data 
</code></pre></div>
<div class="highlight"><pre><span></span><code>  1. exclude_columns
</code></pre></div>
<div class="highlight"><pre><span></span><code>  2. filter_nonproductive
</code></pre></div>
<div class="highlight"><pre><span></span><code>✔ Preprocessing plan is ready
</code></pre></div>
<div class="highlight"><pre><span></span><code>
</code></pre></div>
<div class="highlight"><pre><span></span><code>── Aggregating the data to receptors 
</code></pre></div>
<div class="highlight"><pre><span></span><code>ℹ Found target locus: TRB. The dataset will be pre-filtered to leave chains for this locus only
</code></pre></div>
<div class="highlight"><pre><span></span><code>ℹ Processing data as single-cell sequencing immune repertoires - no counts, with barcodes, chain pairing is possible
</code></pre></div>
<div class="highlight"><pre><span></span><code>✔ Execution plan for receptor data aggregation and annotation is ready
</code></pre></div>
<div class="highlight"><pre><span></span><code>
</code></pre></div>
<div class="highlight"><pre><span></span><code>── Joining the metadata table with the dataset using &#39;filename&#39; column 
</code></pre></div>
<div class="highlight"><pre><span></span><code>✔ Joining plan is ready
</code></pre></div>
<div class="highlight"><pre><span></span><code>
</code></pre></div>
<div class="highlight"><pre><span></span><code>── Postprocessing the data 
</code></pre></div>
<div class="highlight"><pre><span></span><code>  1. prefix_barcodes
</code></pre></div>
<div class="highlight"><pre><span></span><code>✔ Postprocessing plan is ready
</code></pre></div>
<div class="highlight"><pre><span></span><code>
</code></pre></div>
<div class="highlight"><pre><span></span><code>── Saving the newly created ImmunData to disk 
</code></pre></div>
<div class="highlight"><pre><span></span><code>ℹ Writing the receptor annotation data to [/home/runner/work/_temp/Library/immundata/extdata/single_cell/immundata-lb6.csv/annotations.parquet]
</code></pre></div>
<div class="highlight"><pre><span></span><code>ℹ Writing the metadata to [/home/runner/work/_temp/Library/immundata/extdata/single_cell/immundata-lb6.csv/metadata.json]
</code></pre></div>
<div class="highlight"><pre><span></span><code>✔ ImmunData files saved to [/home/runner/work/_temp/Library/immundata/extdata/single_cell/immundata-lb6.csv]
</code></pre></div>
<div class="highlight"><pre><span></span><code>ℹ Reading ImmunData files from [&#39;/home/runner/work/_temp/Library/immundata/extdata/single_cell/immundata-lb6.csv&#39;]
</code></pre></div>
<div class="highlight"><pre><span></span><code>✔ Loaded ImmunData with the receptor schema: [c(&quot;cdr3&quot;, &quot;v_call&quot;) and TRB]
</code></pre></div>
<div class="highlight"><pre><span></span><code>ℹ Reading ImmunData files from [&#39;/home/runner/work/_temp/Library/immundata/extdata/single_cell/immundata-lb6.csv&#39;]
</code></pre></div>
<div class="highlight"><pre><span></span><code>
</code></pre></div>
<div class="highlight"><pre><span></span><code>── Aggregating repertoires... 
</code></pre></div>
<div class="highlight"><pre><span></span><code>✔ Aggregation is finished
</code></pre></div>
<div class="highlight"><pre><span></span><code>
</code></pre></div>
<div class="highlight"><pre><span></span><code>── Summary 
</code></pre></div>
<div class="highlight"><pre><span></span><code>ℹ Time elapsed: 5.77 secs
</code></pre></div>
<div class="highlight"><pre><span></span><code>✔ Loaded ImmunData with the receptor schema: [c(&quot;cdr3&quot;, &quot;v_call&quot;) and TRB]
</code></pre></div>
<div class="highlight"><pre><span></span><code>✔ Loaded ImmunData with the repertoire schema: [Tissue]
</code></pre></div>
<div class="highlight"><pre><span></span><code>✔ Loaded ImmunData with [18714] chains
</code></pre></div>
<div class="cell-code highlight"><pre><span></span><code><span class="kc">NULL</span>
</code></pre></div>
<div class="highlight"><pre><span></span><code>NULL
</code></pre></div>
</div>
</div>
</div>
<p>The result, <code>idata</code>, is an ImmunData object --- a columnar, on-disk
structure similar in spirit to Seurat or AnnData, but specifically
optimized for immune repertoire analysis. You can query, annotate, or
regroup it on the fly, and it remains memory efficient and fully
reproducible.</p>
<h3 id="load-scrnaseq-data">Load scRNAseq data</h3>
<p>If you're working with single-cell transcriptomics, you'll usually want
to bring in cluster annotations or cell type labels. This can be done
with a simple TSV table, or directly from a Seurat/AnnData object.</p>
<div class="tabbed-set tabbed-alternate" data-tabs="10:1"><input checked="checked" id="__tabbed_10_1" name="__tabbed_10" type="radio" /><div class="tabbed-labels"><label for="__tabbed_10_1">R</label></div>
<div class="tabbed-content">
<div class="tabbed-block">
<div class="cell-code highlight"><pre><span></span><code><span class="n">cells_file</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">system.file</span><span class="p">(</span><span class="s">&quot;extdata/single_cell&quot;</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;cells.tsv.gz&quot;</span><span class="p">,</span><span class="w"> </span><span class="n">package</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;immundata&quot;</span><span class="p">)</span>
<span class="n">cells_file</span>
</code></pre></div>
<div class="highlight"><pre><span></span><code>[1] &quot;/home/runner/work/_temp/Library/immundata/extdata/single_cell/cells.tsv.gz&quot;
</code></pre></div>
<div class="cell-code highlight"><pre><span></span><code><span class="n">cells</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">readr</span><span class="o">::</span><span class="nf">read_tsv</span><span class="p">(</span><span class="n">cells_file</span><span class="p">)</span>
</code></pre></div>
<div class="highlight"><pre><span></span><code>Rows: 15174 Columns: 6
── Column specification ────────────────────────────────────────────────────────
Delimiter: &quot;\t&quot;
chr (4): barcode, ident, sample, source
dbl (2): UMAP_1, UMAP_2

ℹ Use `spec()` to retrieve the full column specification for this data.
ℹ Specify the column types or set `show_col_types = FALSE` to quiet this message.
</code></pre></div>
<div class="cell-code highlight"><pre><span></span><code><span class="nf">head</span><span class="p">(</span><span class="n">cells</span><span class="p">)</span>
</code></pre></div>
<div class="highlight"><pre><span></span><code># A tibble: 6 × 6
  barcode                ident     sample source UMAP_1  UMAP_2
  &lt;chr&gt;                  &lt;chr&gt;     &lt;chr&gt;  &lt;chr&gt;   &lt;dbl&gt;   &lt;dbl&gt;
1 LT6_AAACCTGAGCGTTCCG-1 8.4-Chrom LT6    Tumor   0.870  0.0175
2 LT6_AAACCTGCATCGTCGG-1 4.4-FOS   LT6    Tumor  -3.75  -0.445 
3 LT6_AAACCTGGTACGAAAT-1 4.6a-Treg LT6    Tumor  -2.28  -4.70  
4 LT6_AAACCTGGTAGAGTGC-1 4.1-Trm   LT6    Tumor  -3.31  -1.28  
5 LT6_AAACCTGGTTACGTCA-1 4.3-TCF7  LT6    Tumor  -6.93  -0.653 
6 LT6_AAACCTGTCCGTTGCT-1 4.6b-Treg LT6    Tumor  -0.644 -5.03  
</code></pre></div>
</div>
</div>
</div>
<p>To visualize gene expression and link it with immune repertoires, load
the expression matrix into a Seurat object. Here, we use a cached RDS to
speed up the tutorial:</p>
<div class="tabbed-set tabbed-alternate" data-tabs="11:1"><input checked="checked" id="__tabbed_11_1" name="__tabbed_11" type="radio" /><div class="tabbed-labels"><label for="__tabbed_11_1">R</label></div>
<div class="tabbed-content">
<div class="tabbed-block">
<div class="cell-code highlight"><pre><span></span><code><span class="n">url</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="s">&quot;https://zenodo.org/records/15604205/files/l6data.rds?download=1&quot;</span>
<span class="n">dest</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">BiocFileCache</span><span class="o">::</span><span class="nf">bfcrpath</span><span class="p">(</span><span class="n">BiocFileCache</span><span class="o">::</span><span class="nf">BiocFileCache</span><span class="p">(</span><span class="n">ask</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kc">FALSE</span><span class="p">),</span><span class="w"> </span><span class="n">url</span><span class="p">)</span>
</code></pre></div>
<div class="highlight"><pre><span></span><code>adding rname &#39;https://zenodo.org/records/15604205/files/l6data.rds?download=1&#39;
</code></pre></div>
<div class="cell-code highlight"><pre><span></span><code><span class="n">mat</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">readr</span><span class="o">::</span><span class="nf">read_rds</span><span class="p">(</span><span class="n">dest</span><span class="p">)</span>

<span class="n">sdata</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">CreateSeuratObject</span><span class="p">(</span><span class="n">counts</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">mat</span><span class="p">)</span>
</code></pre></div>
<div class="highlight"><pre><span></span><code>Warning: Data is of class matrix. Coercing to dgCMatrix.
</code></pre></div>
<div class="cell-code highlight"><pre><span></span><code><span class="n">embeddings</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">as.matrix</span><span class="p">(</span><span class="n">cells</span><span class="p">[</span><span class="nf">c</span><span class="p">(</span><span class="s">&quot;UMAP_1&quot;</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;UMAP_2&quot;</span><span class="p">)])</span>
<span class="nf">rownames</span><span class="p">(</span><span class="n">embeddings</span><span class="p">)</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">cells</span><span class="o">$</span><span class="n">barcode</span>

<span class="n">umap_dr</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">CreateDimReducObject</span><span class="p">(</span><span class="n">embeddings</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">embeddings</span><span class="p">,</span><span class="w"> </span><span class="n">key</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;UMAP_&quot;</span><span class="p">,</span><span class="w"> </span><span class="n">assay</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">DefaultAssay</span><span class="p">(</span><span class="n">sdata</span><span class="p">))</span>

<span class="n">sdata</span><span class="p">[[</span><span class="s">&quot;umap&quot;</span><span class="p">]]</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">umap_dr</span>

<span class="nf">Idents</span><span class="p">(</span><span class="n">sdata</span><span class="p">)</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">setNames</span><span class="p">(</span><span class="n">cells</span><span class="o">$</span><span class="n">ident</span><span class="p">,</span><span class="w"> </span><span class="n">cells</span><span class="o">$</span><span class="n">barcode</span><span class="p">)</span>

<span class="nf">DimPlot</span><span class="p">(</span><span class="n">sdata</span><span class="p">,</span><span class="w"> </span><span class="n">reduction</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;umap&quot;</span><span class="p">,</span><span class="w"> </span><span class="n">order</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">sort</span><span class="p">(</span><span class="nf">unique</span><span class="p">(</span><span class="n">cells</span><span class="o">$</span><span class="n">ident</span><span class="p">),</span><span class="w"> </span><span class="n">decreasing</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kc">TRUE</span><span class="p">),</span><span class="w"> </span><span class="n">label</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kc">TRUE</span><span class="p">,</span><span class="w"> </span><span class="n">alpha</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">.</span><span class="m">5</span><span class="p">)</span>
</code></pre></div>
<div class="highlight"><pre><span></span><code>Warning: `aes_string()` was deprecated in ggplot2 3.0.0.
ℹ Please use tidy evaluation idioms with `aes()`.
ℹ See also `vignette(&quot;ggplot2-in-packages&quot;)` for more information.
ℹ The deprecated feature was likely used in the Seurat package.
  Please report the issue at &lt;https://github.com/satijalab/seurat/issues&gt;.
</code></pre></div>
<p><img alt="" src="../single_cell_files/figure-markdown/load-dataset-seurat-1.png" /></p>
</div>
</div>
</div>
<blockquote>
<p><strong>Tip:</strong> The <code>cells</code> object here includes precomputed UMAP coordinates
to save time. In a real-world workflow, you would extract cluster
labels and embeddings from your processed Seurat/AnnData object and
merge them with your immune receptor data for downstream analysis.</p>
</blockquote>
<h3 id="how-to-read-other-types-of-data-bulk-single-chain-etc">How to read other types of data --- bulk, single-chain, etc.</h3>
<p>The <code>read_repertoires()</code> function is flexible and supports many
repertoire data types and formats. Here are practical patterns for
different scenarios. Use these examples as templates to adapt to your
own data sources and formats.</p>
<blockquote>
<p><strong>Note:</strong> These code blocks are for demonstration and not intended to
run as part of this tutorial. Please mind the "Tissue": this column
<em>must</em> be present in your data in order to create repertoires by it.</p>
</blockquote>
<h4 id="1-bulk-sequencing">1. Bulk sequencing</h4>
<p>For bulk AIRR-seq data, you generally don't need <code>barcode_col</code>,
<code>locus_col</code>, or <code>umi_col</code>. Optionally provide <code>count_col</code> if your data
includes receptor counts.</p>
<div class="tabbed-set tabbed-alternate" data-tabs="12:1"><input checked="checked" id="__tabbed_12_1" name="__tabbed_12" type="radio" /><div class="tabbed-labels"><label for="__tabbed_12_1">R</label></div>
<div class="tabbed-content">
<div class="tabbed-block">
<div class="cell-code highlight"><pre><span></span><code><span class="n">idata</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">read_repertoires</span><span class="p">(</span><span class="n">path</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">inp_files</span><span class="p">,</span><span class="w"> </span>
<span class="w">                          </span><span class="n">schema</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">schema</span><span class="p">,</span><span class="w"> </span>
<span class="w">                          </span><span class="n">metadata</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">md_table</span><span class="p">,</span><span class="w"> </span>
<span class="w">                          </span><span class="n">count_col</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;counts&quot;</span><span class="p">,</span><span class="w"> </span>
<span class="w">                          </span><span class="n">preprocess</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">make_default_preprocessing</span><span class="p">(</span><span class="s">&quot;airr&quot;</span><span class="p">),</span><span class="w"> </span>
<span class="w">                          </span><span class="n">repertoire_schema</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;Tissue&quot;</span><span class="p">)</span>
</code></pre></div>
</div>
</div>
</div>
<h4 id="2-single-cell-sequencing">2. Single-cell sequencing</h4>
<p>For single-cell data, you'll typically need all columns (barcode, locus,
UMI) for proper chain aggregation and cell assignment.</p>
<div class="tabbed-set tabbed-alternate" data-tabs="13:1"><input checked="checked" id="__tabbed_13_1" name="__tabbed_13" type="radio" /><div class="tabbed-labels"><label for="__tabbed_13_1">R</label></div>
<div class="tabbed-content">
<div class="tabbed-block">
<div class="cell-code highlight"><pre><span></span><code><span class="n">idata</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">read_repertoires</span><span class="p">(</span><span class="n">path</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">inp_files</span><span class="p">,</span><span class="w"> </span>
<span class="w">                          </span><span class="n">schema</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">schema</span><span class="p">,</span><span class="w"> </span>
<span class="w">                          </span><span class="n">metadata</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">md_table</span><span class="p">,</span><span class="w"> </span>
<span class="w">                          </span><span class="n">barcode_col</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;barcode&quot;</span><span class="p">,</span><span class="w"> </span>
<span class="w">                          </span><span class="n">locus_col</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;chain&quot;</span><span class="p">,</span><span class="w"> </span>
<span class="w">                          </span><span class="n">umi_col</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;umis&quot;</span><span class="p">,</span><span class="w"> </span>
<span class="w">                          </span><span class="n">preprocess</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">make_default_preprocessing</span><span class="p">(</span><span class="s">&quot;10x&quot;</span><span class="p">),</span><span class="w"> </span>
<span class="w">                          </span><span class="n">repertoire_schema</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;Tissue&quot;</span><span class="p">)</span>
</code></pre></div>
</div>
</div>
</div>
<h4 id="3-airr-c-or-10x-genomics-format">3. AIRR-C or 10x Genomics format</h4>
<p>Both AIRR-C and 10x Genomics formats are supported. Pay attention to the
receptor schema, locus column (<code>locus</code> or <code>chain</code>), and use the correct
preset for <code>make_default_preprocessing()</code>.</p>
<div class="tabbed-set tabbed-alternate" data-tabs="14:1"><input checked="checked" id="__tabbed_14_1" name="__tabbed_14" type="radio" /><div class="tabbed-labels"><label for="__tabbed_14_1">R</label></div>
<div class="tabbed-content">
<div class="tabbed-block">
<div class="cell-code highlight"><pre><span></span><code><span class="c1">#</span>
<span class="c1"># AIRR-C</span>
<span class="c1">#</span>
<span class="n">schema</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">make_receptor_schema</span><span class="p">(</span><span class="n">features</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="s">&quot;junction_aa&quot;</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;v_call&quot;</span><span class="p">),</span><span class="w"> </span><span class="n">chains</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;TCRB&quot;</span><span class="p">)</span>
<span class="c1"># or</span>
<span class="n">schema</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">make_receptor_schema</span><span class="p">(</span><span class="n">features</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="s">&quot;cdr3_aa&quot;</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;v_call&quot;</span><span class="p">),</span><span class="w"> </span><span class="n">chains</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;TCRB&quot;</span><span class="p">)</span>
<span class="c1"># reminder how to read paired-chain data if needed:</span>
<span class="n">schema</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">make_receptor_schema</span><span class="p">(</span><span class="n">features</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="s">&quot;cdr3_aa&quot;</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;v_call&quot;</span><span class="p">),</span><span class="w"> </span><span class="n">chains</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="s">&quot;TCRA&quot;</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;TCRB&quot;</span><span class="p">))</span>

<span class="n">idata</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">read_repertoires</span><span class="p">(</span><span class="n">path</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">inp_files</span><span class="p">,</span><span class="w"> </span>
<span class="w">                          </span><span class="n">schema</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">schema</span><span class="p">,</span><span class="w"> </span>
<span class="w">                          </span><span class="n">metadata</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">md_table</span><span class="p">,</span>
<span class="w">                          </span><span class="n">barcode_col</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;cell_id&quot;</span><span class="p">,</span>
<span class="w">                          </span><span class="n">locus_col</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;locus&quot;</span><span class="p">,</span>
<span class="w">                          </span><span class="n">umi_col</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;umi_count&quot;</span><span class="p">,</span><span class="w"> </span>
<span class="w">                          </span><span class="n">preprocess</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">make_default_preprocessing</span><span class="p">(</span><span class="s">&quot;airr&quot;</span><span class="p">),</span><span class="w"> </span>
<span class="w">                          </span><span class="n">repertoire_schema</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;Tissue&quot;</span><span class="p">)</span>

<span class="c1">#</span>
<span class="c1"># 10XGenomics</span>
<span class="c1">#</span>

<span class="c1"># This is how the original schema would look like</span>
<span class="n">schema</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">make_receptor_schema</span><span class="p">(</span><span class="n">features</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="s">&quot;cdr3_aa&quot;</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;v_gene&quot;</span><span class="p">),</span><span class="w"> </span><span class="n">chains</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;TRB&quot;</span><span class="p">)</span>

<span class="n">idata</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">read_repertoires</span><span class="p">(</span><span class="n">path</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">inp_files</span><span class="p">,</span><span class="w"> </span>
<span class="w">                          </span><span class="n">schema</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">schema</span><span class="p">,</span><span class="w"> </span>
<span class="w">                          </span><span class="n">metadata</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">md_table</span><span class="p">,</span><span class="w"> </span>
<span class="w">                          </span><span class="n">barcode_col</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;barcode&quot;</span><span class="p">,</span><span class="w"> </span>
<span class="w">                          </span><span class="n">locus_col</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;chain&quot;</span><span class="p">,</span><span class="w"> </span>
<span class="w">                          </span><span class="n">umi_col</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;umis&quot;</span><span class="p">,</span><span class="w"> </span>
<span class="w">                          </span><span class="n">preprocess</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">make_default_preprocessing</span><span class="p">(</span><span class="s">&quot;10x&quot;</span><span class="p">),</span>
<span class="w">                          </span><span class="n">rename_columns</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kc">NULL</span><span class="p">,</span>
<span class="w">                          </span><span class="n">repertoire_schema</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;Tissue&quot;</span><span class="p">)</span>

<span class="c1"># Immundata renames some columns by default to match them with AIRR-C format.</span>
<span class="c1"># So we use this:</span>
<span class="n">schema</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">make_receptor_schema</span><span class="p">(</span><span class="n">features</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="s">&quot;cdr3&quot;</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;v_call&quot;</span><span class="p">),</span><span class="w"> </span><span class="n">chains</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;TRB&quot;</span><span class="p">)</span>

<span class="n">idata</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">read_repertoires</span><span class="p">(</span><span class="n">path</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">inp_files</span><span class="p">,</span><span class="w"> </span>
<span class="w">                          </span><span class="n">schema</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">schema</span><span class="p">,</span><span class="w"> </span>
<span class="w">                          </span><span class="n">metadata</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">md_table</span><span class="p">,</span><span class="w"> </span>
<span class="w">                          </span><span class="n">barcode_col</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;barcode&quot;</span><span class="p">,</span><span class="w"> </span>
<span class="w">                          </span><span class="n">locus_col</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;locus&quot;</span><span class="p">,</span><span class="w"> </span>
<span class="w">                          </span><span class="n">umi_col</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;umis&quot;</span><span class="p">,</span><span class="w"> </span>
<span class="w">                          </span><span class="n">preprocess</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">make_default_preprocessing</span><span class="p">(</span><span class="s">&quot;10x&quot;</span><span class="p">),</span>
<span class="w">                          </span><span class="n">repertoire_schema</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;Tissue&quot;</span><span class="p">)</span>
</code></pre></div>
</div>
</div>
</div>
<h4 id="4-reading-file-paths-from-metadata">4. Reading file paths from metadata</h4>
<p>If your metadata table contains file paths, you can use the special
<code>path = "&lt;metadata&gt;"</code> argument. Specify which column in your metadata
has the file paths with <code>metadata_file_col</code>.</p>
<div class="tabbed-set tabbed-alternate" data-tabs="15:1"><input checked="checked" id="__tabbed_15_1" name="__tabbed_15" type="radio" /><div class="tabbed-labels"><label for="__tabbed_15_1">R</label></div>
<div class="tabbed-content">
<div class="tabbed-block">
<div class="cell-code highlight"><pre><span></span><code><span class="n">idata</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">read_repertoires</span><span class="p">(</span><span class="n">path</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;&lt;metadata&gt;&quot;</span><span class="p">,</span><span class="w"> </span>
<span class="w">                          </span><span class="n">schema</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">schema</span><span class="p">,</span><span class="w"> </span>
<span class="w">                          </span><span class="n">metadata</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">md_table</span><span class="p">,</span><span class="w"> </span>
<span class="w">                          </span><span class="n">barcode_col</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;barcode&quot;</span><span class="p">,</span><span class="w"> </span>
<span class="w">                          </span><span class="n">locus_col</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;locus&quot;</span><span class="p">,</span>
<span class="w">                          </span><span class="n">umi_col</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;umis&quot;</span><span class="p">,</span><span class="w"> </span>
<span class="w">                          </span><span class="n">preprocess</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">make_default_preprocessing</span><span class="p">(</span><span class="s">&quot;10x&quot;</span><span class="p">),</span>
<span class="w">                          </span><span class="n">repertoire_schema</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;Tissue&quot;</span><span class="p">)</span>
</code></pre></div>
</div>
</div>
</div>
<hr />
<h2 id="analyse-immune-repertoires">Analyse immune repertoires</h2>
<h3 id="key-immune-repertoire-statistics">Key immune repertoire statistics</h3>
<p>A well-designed immune repertoire analysis always starts with a set of
basic descriptive statistics. These help you:</p>
<ul>
<li>Check data quality (e.g., are all samples/cell types
    well-represented?)</li>
<li>Spot technical artifacts (e.g., biased V gene usage, length skews)</li>
<li>Summarize diversity and composition for downstream biological or
    clinical questions</li>
</ul>
<p>The <code>immunarch</code> package provide a family of functions, <code>airr_stats_*</code>,
to quickly compute these statistics from your <code>ImmunData</code> object. To see
all options or details, run <code>?airr_stats</code> in your R console.</p>
<p>Currently there are three functions to compute key statistics:</p>
<ol>
<li>
<p><strong><code>airr_stats_chains()</code></strong> <em>Counts the number of chains (e.g., TRA,
    TRB, IGH) per repertoire/sample.</em> Use this for library size, capture
    depth, and to check chain balance (e.g., are you missing TRA?).</p>
</li>
<li>
<p><strong><code>airr_stats_lengths()</code></strong> <em>Summarizes the CDR3 length distribution
    per repertoire.</em> Reveals if your data has technical biases, primer
    effects, or biologically meaningful skews (like more long or short
    clones in a disease).</p>
</li>
<li>
<p><strong><code>airr_stats_genes()</code></strong> <em>Counts the usage of V, D, or J genes per
    repertoire/sample/cluster.</em> Essential for characterizing repertoire
    composition, comparing patient groups, or making features for ML
    models.</p>
</li>
</ol>
<h4 id="1a-number-of-receptors-and-barcodes-for-samples">1a. Number of receptors and barcodes for samples</h4>
<p>Let's start by quantifying two basic properties:</p>
<ul>
<li>
<p>Number of unique barcodes per sample (reflects cell yield /
    diversity)</p>
</li>
<li>
<p>Number of unique receptors per sample (reflects immune diversity and
    capture)</p>
</li>
</ul>
<p>These metrics are often used as a first-line QC and as a way to detect
outliers or sample-specific technical effects.</p>
<div class="tabbed-set tabbed-alternate" data-tabs="16:1"><input checked="checked" id="__tabbed_16_1" name="__tabbed_16" type="radio" /><div class="tabbed-labels"><label for="__tabbed_16_1">R</label></div>
<div class="tabbed-content">
<div class="tabbed-block">
<div class="cell-code highlight"><pre><span></span><code><span class="n">idata_stats</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">airr_stats_chains</span><span class="p">(</span><span class="n">idata</span><span class="p">)</span>

<span class="n">idata_stats</span>
</code></pre></div>
<div class="highlight"><pre><span></span><code># A tibble: 3 × 6
  imd_repertoire_id n_barcodes n_receptors locus n_chains Tissue
*             &lt;int&gt;      &lt;dbl&gt;       &lt;int&gt; &lt;chr&gt;    &lt;int&gt; &lt;chr&gt; 
1                 1       4085        3976 TRB       4085 Blood 
2                 3       7832        3962 TRB       7832 Tumor 
3                 2       6797        2950 TRB       6797 Normal
</code></pre></div>
<div class="cell-code highlight"><pre><span></span><code><span class="n">p1</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">idata_stats</span><span class="w"> </span><span class="o">|&gt;</span><span class="w"> </span><span class="nf">vis</span><span class="p">(</span><span class="n">xval</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;Tissue&quot;</span><span class="p">,</span><span class="w"> </span><span class="n">yval</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;n_receptors&quot;</span><span class="p">,</span><span class="w"> </span><span class="n">fill</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;Tissue&quot;</span><span class="p">,</span><span class="w"> </span><span class="n">title</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;No. barcodes per sample&quot;</span><span class="p">)</span>
<span class="n">p2</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">idata_stats</span><span class="w"> </span><span class="o">|&gt;</span><span class="w"> </span><span class="nf">vis</span><span class="p">(</span><span class="n">xval</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;Tissue&quot;</span><span class="p">,</span><span class="w"> </span><span class="n">yval</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;n_receptors&quot;</span><span class="p">,</span><span class="w"> </span><span class="n">fill</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;Tissue&quot;</span><span class="p">,</span><span class="w"> </span><span class="n">title</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;No. receptors per sample&quot;</span><span class="p">)</span>
<span class="n">p1</span><span class="w"> </span><span class="o">+</span><span class="w"> </span><span class="n">p2</span>
</code></pre></div>
<p><img alt="" src="../single_cell_files/figure-markdown/repertoire-stats-sample-plot-1.png" /></p>
</div>
</div>
</div>
<ul>
<li>
<p>Possible interpretation of results:</p>
<ul>
<li>
<p>Large differences in barcode count may indicate sample quality
    or biological differences (e.g., tumour-infiltrating T cell
    abundance, malignant cells).</p>
</li>
<li>
<p>Receptor count per sample gives a sense of diversity captured
    and can flag technical dropouts or high expansion.</p>
</li>
</ul>
</li>
</ul>
<h4 id="1b-number-of-receptors-and-barcodes-for-cell-clusters">1b. Number of receptors and barcodes for cell clusters</h4>
<p>The basic sample-level view is useful, but often you want to see how
immune repertoires are distributed within cell subpopulations, e.g., by
T cell cluster. With single-cell data, this is straightforward thanks to
the flexible annotation model in <code>immundata</code>.</p>
<p>Currently, the <code>idata</code> object holds three repertoires---one for each
tissue. Let's check that:</p>
<div class="tabbed-set tabbed-alternate" data-tabs="17:1"><input checked="checked" id="__tabbed_17_1" name="__tabbed_17" type="radio" /><div class="tabbed-labels"><label for="__tabbed_17_1">R</label></div>
<div class="tabbed-content">
<div class="tabbed-block">
<div class="cell-code highlight"><pre><span></span><code><span class="n">idata</span><span class="o">$</span><span class="n">repertoires</span>
</code></pre></div>
<div class="highlight"><pre><span></span><code># A tibble: 3 × 4
  imd_repertoire_id Tissue n_barcodes n_receptors
              &lt;int&gt; &lt;chr&gt;       &lt;dbl&gt;       &lt;int&gt;
1                 1 Blood        4085        3976
2                 2 Normal       6797        2950
3                 3 Tumor        7832        3962
</code></pre></div>
</div>
</div>
</div>
<p>To break down repertoire statistics by cluster, we'll annotate immune
receptors with cluster labels from the single-cell data (<code>sdata</code>). Both
<code>idata</code> and <code>cells</code> (and the Seurat object) share cell barcodes, so we
can safely map clusters onto the immune data:</p>
<div class="tabbed-set tabbed-alternate" data-tabs="18:1"><input checked="checked" id="__tabbed_18_1" name="__tabbed_18" type="radio" /><div class="tabbed-labels"><label for="__tabbed_18_1">R</label></div>
<div class="tabbed-content">
<div class="tabbed-block">
<div class="cell-code highlight"><pre><span></span><code><span class="c1"># Build a tibble: cluster label per barcode (cell)</span>
<span class="n">annot</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">tibble</span><span class="p">(</span><span class="n">Cluster</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">as.character</span><span class="p">(</span><span class="nf">Idents</span><span class="p">(</span><span class="n">sdata</span><span class="p">)),</span><span class="w"> </span><span class="n">barcode</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">names</span><span class="p">(</span><span class="nf">Idents</span><span class="p">(</span><span class="n">sdata</span><span class="p">)))</span>

<span class="n">annot</span>
</code></pre></div>
<div class="highlight"><pre><span></span><code># A tibble: 15,174 × 2
   Cluster   barcode               
   &lt;chr&gt;     &lt;chr&gt;                 
 1 8.4-Chrom LT6_AAACCTGAGCGTTCCG-1
 2 4.4-FOS   LT6_AAACCTGCATCGTCGG-1
 3 4.6a-Treg LT6_AAACCTGGTACGAAAT-1
 4 4.1-Trm   LT6_AAACCTGGTAGAGTGC-1
 5 4.3-TCF7  LT6_AAACCTGGTTACGTCA-1
 6 4.6b-Treg LT6_AAACCTGTCCGTTGCT-1
 7 8.2-Tem   LT6_AAACCTGTCGCCTGTT-1
 8 4.1-Trm   LT6_AAACCTGTCTCCGGTT-1
 9 4.6a-Treg LT6_AAACGGGAGGTAGCTG-1
10 8.3a-Trm  LT6_AAACGGGAGTTATCGC-1
# ℹ 15,164 more rows
</code></pre></div>
<div class="cell-code highlight"><pre><span></span><code><span class="c1"># Annotate ImmunData by barcode, then regroup repertoires by Tissue *and* Cluster</span>
<span class="n">idata</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">annotate_barcodes</span><span class="p">(</span><span class="n">idata</span><span class="p">,</span><span class="w"> </span><span class="n">annot</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;barcode&quot;</span><span class="p">)</span>
<span class="n">idata</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">agg_repertoires</span><span class="p">(</span><span class="n">idata</span><span class="p">,</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="s">&quot;Tissue&quot;</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;Cluster&quot;</span><span class="p">))</span>

<span class="c1"># Now each repertoire is a unique (Tissue, Cluster) pair</span>
<span class="n">idata</span><span class="o">$</span><span class="n">repertoires</span>
</code></pre></div>
<div class="highlight"><pre><span></span><code># A tibble: 51 × 5
   imd_repertoire_id Tissue Cluster     n_barcodes n_receptors
               &lt;int&gt; &lt;chr&gt;  &lt;chr&gt;            &lt;dbl&gt;       &lt;int&gt;
 1                 1 Blood  8.2-Tem            134         123
 2                 2 Blood  4.3-TCF7          1054        1049
 3                 3 Blood  4.6a-Treg          208         203
 4                 4 Blood  8.3c-Trm            53          53
 5                 5 Blood  8.6-KLRB1           75          71
 6                 6 Blood  8.5-Mitosis         43          42
 7                 7 Blood  8.3b-Trm            32          31
 8                 8 Normal 3.1-MT             248         183
 9                 9 Normal &lt;NA&gt;               680         420
10                10 Tumor  3.1-MT              93          74
# ℹ 41 more rows
</code></pre></div>
</div>
</div>
</div>
<p>Now is a good time to save a snapshot. You've run several heavy steps
and defined your working dataset: tissue annotations, cluster labels,
and more. A snapshot freezes this state for reproducibility. Learn why
in the <a href="https://immunomind.github.io/docs/concepts/basics/immutability/">Concepts →
Immutability</a>.</p>
<p>You can create snapshot via <code>write_immundata</code> function that will save
the data on disk, and then load the snapshot so you can work with the
newly created data.</p>
<div class="tabbed-set tabbed-alternate" data-tabs="19:1"><input checked="checked" id="__tabbed_19_1" name="__tabbed_19" type="radio" /><div class="tabbed-labels"><label for="__tabbed_19_1">R</label></div>
<div class="tabbed-content">
<div class="tabbed-block">
<div class="cell-code highlight"><pre><span></span><code><span class="n">idata</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">write_immundata</span><span class="p">(</span><span class="n">idata</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;./immundata-files&quot;</span><span class="p">)</span>
</code></pre></div>
<div class="highlight"><pre><span></span><code>ℹ Writing the receptor annotation data to [./immundata-files/annotations.parquet]
</code></pre></div>
<div class="highlight"><pre><span></span><code>ℹ Writing the metadata to [./immundata-files/metadata.json]
</code></pre></div>
<div class="highlight"><pre><span></span><code>✔ ImmunData files saved to [./immundata-files]
</code></pre></div>
<div class="highlight"><pre><span></span><code>ℹ Reading ImmunData files from [&#39;./immundata-files&#39;]
</code></pre></div>
<div class="highlight"><pre><span></span><code>✔ Loaded ImmunData with the receptor schema: [c(&quot;cdr3&quot;, &quot;v_call&quot;) and TRB]
</code></pre></div>
<div class="highlight"><pre><span></span><code>✔ Loaded ImmunData with the repertoire schema: [Tissue and Cluster]
</code></pre></div>
</div>
</div>
</div>
<p>Now let's recompute and visualize statistics at this new level of
granularity:</p>
<div class="tabbed-set tabbed-alternate" data-tabs="20:1"><input checked="checked" id="__tabbed_20_1" name="__tabbed_20" type="radio" /><div class="tabbed-labels"><label for="__tabbed_20_1">R</label></div>
<div class="tabbed-content">
<div class="tabbed-block">
<div class="cell-code highlight"><pre><span></span><code><span class="n">idata_stats</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">airr_stats_chains</span><span class="p">(</span><span class="n">idata</span><span class="p">)</span>

<span class="n">idata_stats</span>
</code></pre></div>
<div class="highlight"><pre><span></span><code># A tibble: 51 × 7
   imd_repertoire_id n_barcodes n_receptors locus n_chains Tissue Cluster    
 *             &lt;int&gt;      &lt;dbl&gt;       &lt;int&gt; &lt;chr&gt;    &lt;int&gt; &lt;chr&gt;  &lt;chr&gt;      
 1                 1        134         123 TRB        134 Blood  8.2-Tem    
 2                 2       1054        1049 TRB       1054 Blood  4.3-TCF7   
 3                 3        208         203 TRB        208 Blood  4.6a-Treg  
 4                 4         53          53 TRB         53 Blood  8.3c-Trm   
 5                 5         75          71 TRB         75 Blood  8.6-KLRB1  
 6                 6         43          42 TRB         43 Blood  8.5-Mitosis
 7                 7         32          31 TRB         32 Blood  8.3b-Trm   
 8                 8        248         183 TRB        248 Normal 3.1-MT     
 9                 9        680         420 TRB        680 Normal &lt;NA&gt;       
10                10         93          74 TRB         93 Tumor  3.1-MT     
# ℹ 41 more rows
</code></pre></div>
<div class="cell-code highlight"><pre><span></span><code><span class="c1"># Barcodes and receptors per tissue, filled by cluster</span>
<span class="n">p1</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">idata_stats</span><span class="w"> </span><span class="o">|&gt;</span><span class="w"> </span><span class="nf">vis</span><span class="p">(</span><span class="n">xval</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;Tissue&quot;</span><span class="p">,</span><span class="w"> </span><span class="n">yval</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;n_receptors&quot;</span><span class="p">,</span><span class="w"> </span><span class="n">fill</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;Cluster&quot;</span><span class="p">,</span><span class="w"> </span><span class="n">title</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;No. barcodes per sample&quot;</span><span class="p">)</span>
<span class="n">p2</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">idata_stats</span><span class="w"> </span><span class="o">|&gt;</span><span class="w"> </span><span class="nf">vis</span><span class="p">(</span><span class="n">xval</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;Tissue&quot;</span><span class="p">,</span><span class="w"> </span><span class="n">yval</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;n_receptors&quot;</span><span class="p">,</span><span class="w"> </span><span class="n">fill</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;Cluster&quot;</span><span class="p">,</span><span class="w"> </span><span class="n">title</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;No. receptors per sample&quot;</span><span class="p">)</span>
<span class="n">p1</span><span class="w"> </span><span class="o">/</span><span class="w"> </span><span class="n">p2</span>
</code></pre></div>
<p><img alt="" src="../single_cell_files/figure-markdown/repertoire-stats-cluster-plot-1.png" /></p>
</div>
</div>
</div>
<p>This provides a detailed view of which clusters dominate each tissue. To
focus even further on cluster differences across tissues, group and plot
by cluster:</p>
<div class="tabbed-set tabbed-alternate" data-tabs="21:1"><input checked="checked" id="__tabbed_21_1" name="__tabbed_21" type="radio" /><div class="tabbed-labels"><label for="__tabbed_21_1">R</label></div>
<div class="tabbed-content">
<div class="tabbed-block">
<div class="cell-code highlight"><pre><span></span><code><span class="n">idata_stats</span><span class="w"> </span><span class="o">|&gt;</span><span class="w"> </span><span class="nf">vis</span><span class="p">(</span><span class="n">xval</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;Cluster&quot;</span><span class="p">,</span><span class="w"> </span><span class="n">fill</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;Tissue&quot;</span><span class="p">)</span>
</code></pre></div>
<p><img alt="" src="../single_cell_files/figure-markdown/repertoire-stats-cluster-plot-patients-1.png" /></p>
</div>
</div>
</div>
<blockquote>
<p><strong>Note:</strong> You may see NAs, or an unexpectedly high number of receptors
per barcode (especially in tumor samples), if you analyze only the
"TRB" chain. This happens because a single barcode can have multiple
TRB contigs, due to technical or biological reasons (e.g., malignant
cells). Switching to a paired-chain (TRA + TRB) schema forces
<code>immundata</code> to collapse these into a single receptor pair per barcode,
producing a clearer one-cell-one-receptor relationship and making
cluster-level summaries more interpretable.</p>
</blockquote>
<p>Homework for you: try running this analysis with a paired-chain schema
(see earlier examples). The rest of your code will remain unchanged ---
that's the power of decoupling data structure (<code>immundata</code> data
structure) from analysis logic (<code>immunarch</code> functions)!</p>
<p>We can also normalize by the total number of barcodes per tissue for
fairer cross-tissue comparison:</p>
<div class="tabbed-set tabbed-alternate" data-tabs="22:1"><input checked="checked" id="__tabbed_22_1" name="__tabbed_22" type="radio" /><div class="tabbed-labels"><label for="__tabbed_22_1">R</label></div>
<div class="tabbed-content">
<div class="tabbed-block">
<div class="cell-code highlight"><pre><span></span><code><span class="n">idata_stats2</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">idata_stats</span><span class="w"> </span><span class="o">|&gt;</span><span class="w"> </span><span class="nf">mutate</span><span class="p">(</span><span class="n">.by</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">Tissue</span><span class="p">,</span><span class="w"> </span><span class="n">TissueSize</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">sum</span><span class="p">(</span><span class="n">n_barcodes</span><span class="p">))</span><span class="w"> </span><span class="o">|&gt;</span><span class="w"> </span><span class="nf">mutate</span><span class="p">(</span><span class="n">n_barcodes_div</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">n_barcodes</span><span class="w"> </span><span class="o">/</span><span class="w"> </span><span class="n">TissueSize</span><span class="p">)</span>
<span class="n">idata_stats2</span>
</code></pre></div>
<div class="highlight"><pre><span></span><code># A tibble: 51 × 9
   imd_repertoire_id n_barcodes n_receptors locus n_chains Tissue Cluster    
               &lt;int&gt;      &lt;dbl&gt;       &lt;int&gt; &lt;chr&gt;    &lt;int&gt; &lt;chr&gt;  &lt;chr&gt;      
 1                 1        134         123 TRB        134 Blood  8.2-Tem    
 2                 2       1054        1049 TRB       1054 Blood  4.3-TCF7   
 3                 3        208         203 TRB        208 Blood  4.6a-Treg  
 4                 4         53          53 TRB         53 Blood  8.3c-Trm   
 5                 5         75          71 TRB         75 Blood  8.6-KLRB1  
 6                 6         43          42 TRB         43 Blood  8.5-Mitosis
 7                 7         32          31 TRB         32 Blood  8.3b-Trm   
 8                 8        248         183 TRB        248 Normal 3.1-MT     
 9                 9        680         420 TRB        680 Normal &lt;NA&gt;       
10                10         93          74 TRB         93 Tumor  3.1-MT     
# ℹ 41 more rows
# ℹ 2 more variables: TissueSize &lt;dbl&gt;, n_barcodes_div &lt;dbl&gt;
</code></pre></div>
<div class="cell-code highlight"><pre><span></span><code><span class="n">idata_stats2</span><span class="w"> </span><span class="o">|&gt;</span><span class="w"> </span><span class="nf">vis</span><span class="p">(</span><span class="n">xval</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;Cluster&quot;</span><span class="p">,</span><span class="w"> </span><span class="n">y</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;n_barcodes_div&quot;</span><span class="p">,</span><span class="w"> </span><span class="n">fill</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;Tissue&quot;</span><span class="p">,</span><span class="w"> </span><span class="n">title</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;No. barcodes per sample&quot;</span><span class="p">)</span>
</code></pre></div>
<p><img alt="" src="../single_cell_files/figure-markdown/repertoire-stats-cluster-plot-norm-1.png" /></p>
</div>
</div>
</div>
<ul>
<li>
<p>Possible interpretation:</p>
<ul>
<li>
<p>You may observe patterns like more Tregs in tumor, more Trm in
    NAT, etc.</p>
</li>
<li>
<p>Remember: normalization lets you compare proportions across
    samples, but absolute numbers (e.g., total cells per tissue) are
    influenced by capture and sequencing depth.</p>
</li>
</ul>
</li>
</ul>
<h4 id="2a-global-gene-usage">2a. Global gene usage</h4>
<p>One of the most informative immune repertoire features is V(D)J gene
segment usage. Shifts in gene usage can reveal biological signatures
(e.g., enrichment of certain V genes in tumor-infiltrating clones),
technical artifacts, or repertoire biases. Let's create a heatmap
showing the frequency of each V gene segment across all cluster-tissue
combinations. This gives a compact, bird's-eye view of usage patterns
and potential "hotspots" of expansion.</p>
<div class="tabbed-set tabbed-alternate" data-tabs="23:1"><input checked="checked" id="__tabbed_23_1" name="__tabbed_23" type="radio" /><div class="tabbed-labels"><label for="__tabbed_23_1">R</label></div>
<div class="tabbed-content">
<div class="tabbed-block">
<div class="cell-code highlight"><pre><span></span><code><span class="n">gene_usage_full</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">airr_stats_genes</span><span class="p">(</span>
<span class="w">  </span><span class="n">idata</span><span class="p">,</span>
<span class="w">  </span><span class="n">gene_col</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;v_call&quot;</span><span class="p">,</span>
<span class="w">  </span><span class="n">level</span><span class="w">    </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;receptor&quot;</span>
<span class="p">)</span>

<span class="n">gene_usage_full</span><span class="w"> </span><span class="o">|&gt;</span><span class="w"> </span><span class="nf">vis</span><span class="p">(</span><span class="n">col</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="s">&quot;Cluster&quot;</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;Tissue&quot;</span><span class="p">))</span>
</code></pre></div>
<p><img alt="" src="../single_cell_files/figure-markdown/repertoire-gene-usage-plot-1.png" /></p>
</div>
</div>
</div>
<h4 id="2b-most-overrepresented-v-gene-segments">2b. Most overrepresented V gene segments</h4>
<p>As you can see, the broad overview is not very straightforward to
understand. We can focus on most strong "signals" from the data: which V
genes are most used across all samples and repertoires? This can help
spot outlier samples, overall biases, or dominant clonotypes.</p>
<div class="tabbed-set tabbed-alternate" data-tabs="24:1"><input checked="checked" id="__tabbed_24_1" name="__tabbed_24" type="radio" /><div class="tabbed-labels"><label for="__tabbed_24_1">R</label></div>
<div class="tabbed-content">
<div class="tabbed-block">
<div class="cell-code highlight"><pre><span></span><code><span class="c1"># Compute V gene usage for all repertoires (by receptor count, split by tissue if present)</span>
<span class="n">gene_usage_full</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">airr_stats_genes</span><span class="p">(</span>
<span class="w">  </span><span class="n">idata</span><span class="p">,</span>
<span class="w">  </span><span class="n">gene_col</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;v_call&quot;</span><span class="p">,</span>
<span class="w">  </span><span class="n">level</span><span class="w">    </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;receptor&quot;</span>
<span class="p">)</span>

<span class="n">top_n</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="m">10</span>

<span class="n">top_vgenes</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">gene_usage_full</span><span class="w"> </span><span class="o">|&gt;</span><span class="w"> </span>
<span class="w">  </span><span class="nf">group_by</span><span class="p">(</span><span class="n">Tissue</span><span class="p">,</span><span class="w"> </span><span class="n">v_call</span><span class="p">)</span><span class="w"> </span><span class="o">|&gt;</span><span class="w"> </span>
<span class="w">  </span><span class="nf">summarize</span><span class="p">(</span><span class="n">n</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">sum</span><span class="p">(</span><span class="n">n</span><span class="p">),</span><span class="w"> </span><span class="n">.groups</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;drop&quot;</span><span class="p">)</span><span class="w"> </span><span class="o">|&gt;</span><span class="w"> </span>
<span class="w">  </span><span class="nf">group_by</span><span class="p">(</span><span class="n">Tissue</span><span class="p">)</span><span class="w"> </span><span class="o">|&gt;</span><span class="w"> </span>
<span class="w">  </span><span class="nf">slice_max</span><span class="p">(</span><span class="n">order_by</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">n</span><span class="p">,</span><span class="w"> </span><span class="n">n</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">top_n</span><span class="p">)</span><span class="w"> </span><span class="o">|&gt;</span><span class="w"> </span>
<span class="w">  </span><span class="nf">ungroup</span><span class="p">()</span>

<span class="c1"># Collect the *union* of all top genes across tissues</span>
<span class="n">top_gene_set</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">unique</span><span class="p">(</span><span class="n">top_vgenes</span><span class="o">$</span><span class="n">v_call</span><span class="p">)</span>

<span class="c1"># Filter to only these V genes for the heatmap (across all clusters and tissues)</span>
<span class="n">usage_focus</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">gene_usage_full</span><span class="w"> </span><span class="o">|&gt;</span><span class="w"> </span>
<span class="w">  </span><span class="nf">filter</span><span class="p">(</span><span class="n">v_call</span><span class="w"> </span><span class="o">%in%</span><span class="w"> </span><span class="n">top_gene_set</span><span class="p">)</span>

<span class="n">usage_focus</span><span class="w"> </span><span class="o">|&gt;</span><span class="w"> </span><span class="nf">vis</span><span class="p">(</span><span class="n">col</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="s">&quot;Cluster&quot;</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;Tissue&quot;</span><span class="p">))</span>
</code></pre></div>
<p><img alt="" src="../single_cell_files/figure-markdown/repertoire-gene-usage-top-1.png" /></p>
</div>
</div>
</div>
<h4 id="2c-cluster-specific-v-gene-distribution">2c. Cluster-specific V gene distribution</h4>
<p>We can do pretty much the same if there are specific clusters we are
interested in specifically:</p>
<div class="tabbed-set tabbed-alternate" data-tabs="25:1"><input checked="checked" id="__tabbed_25_1" name="__tabbed_25" type="radio" /><div class="tabbed-labels"><label for="__tabbed_25_1">R</label></div>
<div class="tabbed-content">
<div class="tabbed-block">
<div class="cell-code highlight"><pre><span></span><code><span class="c1"># We use special functions from duckdb to filter clusters of interest</span>
<span class="n">gene_usage_full</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">idata</span><span class="w"> </span><span class="o">|&gt;</span><span class="w"> </span>
<span class="w">  </span><span class="nf">filter_immundata</span><span class="p">(</span><span class="n">dd</span><span class="o">$</span><span class="nf">regexp_matches</span><span class="p">(</span><span class="n">Cluster</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;Treg|Trm&quot;</span><span class="p">))</span><span class="w"> </span><span class="o">|&gt;</span><span class="w"> </span>
<span class="w">  </span><span class="nf">airr_stats_genes</span><span class="p">(</span>
<span class="w">    </span><span class="n">gene_col</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;v_call&quot;</span><span class="p">,</span>
<span class="w">    </span><span class="n">level</span><span class="w">    </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;receptor&quot;</span>
<span class="w">  </span><span class="p">)</span>

<span class="n">top_n</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="m">10</span>

<span class="n">top_vgenes</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">gene_usage_full</span><span class="w"> </span><span class="o">|&gt;</span><span class="w"> </span>
<span class="w">  </span><span class="nf">group_by</span><span class="p">(</span><span class="n">Tissue</span><span class="p">,</span><span class="w"> </span><span class="n">v_call</span><span class="p">)</span><span class="w"> </span><span class="o">|&gt;</span><span class="w"> </span>
<span class="w">  </span><span class="nf">summarize</span><span class="p">(</span><span class="n">n</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">sum</span><span class="p">(</span><span class="n">n</span><span class="p">),</span><span class="w"> </span><span class="n">.groups</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;drop&quot;</span><span class="p">)</span><span class="w"> </span><span class="o">|&gt;</span><span class="w"> </span>
<span class="w">  </span><span class="nf">group_by</span><span class="p">(</span><span class="n">Tissue</span><span class="p">)</span><span class="w"> </span><span class="o">|&gt;</span><span class="w"> </span>
<span class="w">  </span><span class="nf">slice_max</span><span class="p">(</span><span class="n">order_by</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">n</span><span class="p">,</span><span class="w"> </span><span class="n">n</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">top_n</span><span class="p">)</span><span class="w"> </span><span class="o">|&gt;</span><span class="w"> </span>
<span class="w">  </span><span class="nf">ungroup</span><span class="p">()</span>

<span class="c1"># Collect the *union* of all top genes across tissues</span>
<span class="n">top_gene_set</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">unique</span><span class="p">(</span><span class="n">top_vgenes</span><span class="o">$</span><span class="n">v_call</span><span class="p">)</span>

<span class="c1"># Filter to only these V genes for the heatmap (across all clusters and tissues)</span>
<span class="n">usage_focus</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">gene_usage_full</span><span class="w"> </span><span class="o">|&gt;</span><span class="w"> </span>
<span class="w">  </span><span class="nf">filter</span><span class="p">(</span><span class="n">v_call</span><span class="w"> </span><span class="o">%in%</span><span class="w"> </span><span class="n">top_gene_set</span><span class="p">)</span>

<span class="n">usage_focus</span><span class="w"> </span><span class="o">|&gt;</span><span class="w"> </span><span class="nf">vis</span><span class="p">(</span><span class="n">col</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="s">&quot;Cluster&quot;</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;Tissue&quot;</span><span class="p">))</span>
</code></pre></div>
<p><img alt="" src="../single_cell_files/figure-markdown/repertoire-gene-usage-cluster-1.png" /></p>
</div>
</div>
</div>
<ul>
<li>
<p>The rows are the most frequent V genes (union of top N per tissue).</p>
</li>
<li>
<p>The columns represent every observed (Cluster, Tissue) combination.</p>
</li>
<li>
<p>Color shows how often each V gene appears in each group.</p>
</li>
</ul>
<h4 id="3a-cdr3-length-distribution-for-the-full-dataset">3a. CDR3 length distribution for the full dataset</h4>
<p>The CDR3 length distribution is a key QC and biological fingerprint of
an immune repertoire. It can reveal technical artifacts (e.g.,
primer/UMI bias), repertoire selection, or expansion of unusual clones
in disease states.</p>
<p>We can compute and visualize CDR3 length distributions across all
tissues and clusters.</p>
<div class="tabbed-set tabbed-alternate" data-tabs="26:1"><input checked="checked" id="__tabbed_26_1" name="__tabbed_26" type="radio" /><div class="tabbed-labels"><label for="__tabbed_26_1">R</label></div>
<div class="tabbed-content">
<div class="tabbed-block">
<div class="cell-code highlight"><pre><span></span><code><span class="c1"># Compute CDR3 length stats for all repertoires (default is cdr3_aa)</span>
<span class="n">length_stats</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">airr_stats_lengths</span><span class="p">(</span>
<span class="w">  </span><span class="n">idata</span><span class="p">,</span>
<span class="w">  </span><span class="n">seq_col</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;cdr3&quot;</span>
<span class="p">)</span>

<span class="c1"># Plot: density of CDR3 lengths by tissue</span>
<span class="nf">ggplot</span><span class="p">(</span><span class="n">length_stats</span><span class="p">,</span><span class="w"> </span><span class="nf">aes</span><span class="p">(</span><span class="n">x</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">seq_len</span><span class="p">,</span><span class="w"> </span><span class="n">y</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">pct</span><span class="p">,</span><span class="w"> </span><span class="n">color</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">Tissue</span><span class="p">))</span><span class="w"> </span><span class="o">+</span>
<span class="w">  </span><span class="nf">geom_line</span><span class="p">(</span><span class="n">stat</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;identity&quot;</span><span class="p">,</span><span class="w"> </span><span class="n">linewidth</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">1.2</span><span class="p">)</span><span class="w"> </span><span class="o">+</span>
<span class="w">  </span><span class="nf">facet_wrap</span><span class="p">(</span><span class="o">~</span><span class="w"> </span><span class="n">Tissue</span><span class="p">,</span><span class="w"> </span><span class="n">scales</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;free_y&quot;</span><span class="p">)</span><span class="w"> </span><span class="o">+</span>
<span class="w">  </span><span class="nf">ggtitle</span><span class="p">(</span><span class="s">&quot;CDR3 length distribution by tissue&quot;</span><span class="p">)</span><span class="w"> </span><span class="o">+</span>
<span class="w">  </span><span class="nf">xlab</span><span class="p">(</span><span class="s">&quot;CDR3 length (amino acids)&quot;</span><span class="p">)</span><span class="w"> </span><span class="o">+</span>
<span class="w">  </span><span class="nf">ylab</span><span class="p">(</span><span class="s">&quot;% receptors&quot;</span><span class="p">)</span><span class="w"> </span><span class="o">+</span>
<span class="w">  </span><span class="n">ggsci</span><span class="o">::</span><span class="nf">scale_color_locuszoom</span><span class="p">()</span>
</code></pre></div>
<p><img alt="" src="../single_cell_files/figure-markdown/repertoire-length-dist-all-1.png" /></p>
</div>
</div>
</div>
<h4 id="3b-cdr3-length-distribution-for-the-selected-clusters">3b. CDR3 length distribution for the selected clusters</h4>
<p>You can also zoom in on clusters of interest (e.g., Trm, Treg) for
fine-grained analysis.</p>
<div class="tabbed-set tabbed-alternate" data-tabs="27:1"><input checked="checked" id="__tabbed_27_1" name="__tabbed_27" type="radio" /><div class="tabbed-labels"><label for="__tabbed_27_1">R</label></div>
<div class="tabbed-content">
<div class="tabbed-block">
<div class="cell-code highlight"><pre><span></span><code><span class="c1"># Choose clusters to display</span>
<span class="n">selected_clusters</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="s">&quot;Trm&quot;</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;Treg&quot;</span><span class="p">)</span>

<span class="n">length_stats_sel</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">length_stats</span><span class="w"> </span><span class="o">%&gt;%</span>
<span class="w">  </span><span class="nf">filter</span><span class="p">(</span><span class="nf">grepl</span><span class="p">(</span><span class="s">&quot;Trm|Treg&quot;</span><span class="p">,</span><span class="w"> </span><span class="n">Cluster</span><span class="p">))</span>

<span class="nf">ggplot</span><span class="p">(</span><span class="n">length_stats_sel</span><span class="p">,</span><span class="w"> </span><span class="nf">aes</span><span class="p">(</span><span class="n">x</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">seq_len</span><span class="p">,</span><span class="w"> </span><span class="n">y</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">pct</span><span class="p">,</span><span class="w"> </span><span class="n">color</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">Tissue</span><span class="p">))</span><span class="w"> </span><span class="o">+</span>
<span class="w">  </span><span class="nf">geom_line</span><span class="p">(</span><span class="n">stat</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;identity&quot;</span><span class="p">,</span><span class="w"> </span><span class="n">linewidth</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">1.2</span><span class="p">)</span><span class="w"> </span><span class="o">+</span>
<span class="w">  </span><span class="nf">facet_wrap</span><span class="p">(</span><span class="o">~</span><span class="w"> </span><span class="n">Cluster</span><span class="p">,</span><span class="w"> </span><span class="n">scales</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;free_y&quot;</span><span class="p">)</span><span class="w"> </span><span class="o">+</span>
<span class="w">  </span><span class="nf">ggtitle</span><span class="p">(</span><span class="s">&quot;CDR3 length distribution in Trm and Treg clusters&quot;</span><span class="p">)</span><span class="w"> </span><span class="o">+</span>
<span class="w">  </span><span class="nf">xlab</span><span class="p">(</span><span class="s">&quot;CDR3 length (amino acids)&quot;</span><span class="p">)</span><span class="w"> </span><span class="o">+</span>
<span class="w">  </span><span class="nf">ylab</span><span class="p">(</span><span class="s">&quot;% receptors&quot;</span><span class="p">)</span><span class="w"> </span><span class="o">+</span>
<span class="w">  </span><span class="n">ggsci</span><span class="o">::</span><span class="nf">scale_color_locuszoom</span><span class="p">()</span>
</code></pre></div>
<p><img alt="" src="../single_cell_files/figure-markdown/repertoire-length-dist-clusters-1.png" /></p>
</div>
</div>
</div>
<ul>
<li>
<p><strong>Possible interpretation:</strong></p>
<ul>
<li>Most TCR/BCR repertoires peak at 13--15 amino acids for CDR3,
    but expansions, artifacts, or certain subsets may show longer or
    shorter tails.</li>
<li>Comparing distributions across tissues/clusters can highlight
    biologically relevant skewing (e.g., enrichment of unusually
    long or short CDR3s in tumors).</li>
</ul>
</li>
</ul>
<h3 id="clonality-quantifying-immune-expansion">Clonality: quantifying immune expansion</h3>
<p>Immune repertoires are rarely flat. Certain T or B cell receptors expand
dramatically in response to infection, cancer, or autoimmunity.
Quantifying clonality lets you:</p>
<ul>
<li>Detect immune responses ("public" expansions or tumor-infiltrating
    clones)</li>
<li>Compare the expansion across samples, tissues, or cell states</li>
<li>Identify technical problems (e.g., overamplification)</li>
</ul>
<p>Currently there are three ways to summarize clonality:</p>
<ol>
<li><strong><code>airr_clonality_line</code></strong> <em>Rank--abundance plots</em> Shows how steep
    the line of receptor abundances is.</li>
<li><strong><code>airr_clonality_prop</code></strong> <em>Clonal space partitions by proportion</em>
    Shows how much space is occupied by hyperexpanded, large, or rare
    clones.</li>
<li><strong><code>airr_clonality_rank</code></strong> <em>Clonal space partitions by rank</em> Shows
    how much space is occupied by receptors from different bins: top-10
    most abundance, top-100 most abundant, etc.</li>
</ol>
<blockquote>
<p><strong>Note:</strong> These functions are repertoire-level statistics. For
annotating <code>ImmunData</code> with the per-receptor information of clonality,
e.g., which receptors are hyperexpanded, to visualize them later on
UMAP, use the receptor-level function family <code>annotate_clonality</code>. We
will discuss them in the sections below.</p>
</blockquote>
<h4 id="1-rank-abundance">1. Rank abundance</h4>
<p>A rank abundance plot shows, for each sample (or group), how cell counts
drop as you go from the most to least abundant clone. In other words,
how steep the clonal hierarchy is.</p>
<ul>
<li><strong>Steep curves</strong>: few highly expanded clones dominate (e.g., tumor
    or antigen-driven response)</li>
<li><strong>Flat curves</strong>: repertoire is diverse and even</li>
</ul>
<div class="tabbed-set tabbed-alternate" data-tabs="28:1"><input checked="checked" id="__tabbed_28_1" name="__tabbed_28" type="radio" /><div class="tabbed-labels"><label for="__tabbed_28_1">R</label></div>
<div class="tabbed-content">
<div class="tabbed-block">
<div class="cell-code highlight"><pre><span></span><code><span class="c1"># Compute the abundance (cell/UMI count) for top 1000 clones per repertoire</span>
<span class="n">clonal_line</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">airr_clonality_line</span><span class="p">(</span>
<span class="w">  </span><span class="n">idata</span><span class="p">,</span>
<span class="w">  </span><span class="n">limit</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">1000</span>
<span class="p">)</span>

<span class="c1"># Plot: each line is a sample, colored by tissue</span>
<span class="nf">ggplot</span><span class="p">(</span><span class="n">clonal_line</span><span class="p">,</span><span class="w"> </span><span class="nf">aes</span><span class="p">(</span><span class="n">x</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">index</span><span class="p">,</span><span class="w"> </span><span class="n">y</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">imd_count</span><span class="p">,</span><span class="w"> </span><span class="n">color</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">Tissue</span><span class="p">,</span><span class="w"> </span><span class="n">group</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">imd_repertoire_id</span><span class="p">))</span><span class="w"> </span><span class="o">+</span>
<span class="w">  </span><span class="nf">geom_line</span><span class="p">(</span><span class="n">alpha</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">0.7</span><span class="p">)</span><span class="w"> </span><span class="o">+</span>
<span class="w">  </span><span class="nf">scale_x_log10</span><span class="p">()</span><span class="w"> </span><span class="o">+</span>
<span class="w">  </span><span class="nf">scale_y_log10</span><span class="p">()</span><span class="w"> </span><span class="o">+</span>
<span class="w">  </span><span class="nf">ggtitle</span><span class="p">(</span><span class="s">&quot;Rank–abundance: Top 1000 clones per repertoire&quot;</span><span class="p">)</span><span class="w"> </span><span class="o">+</span>
<span class="w">  </span><span class="nf">xlab</span><span class="p">(</span><span class="s">&quot;Clone rank (1 = most abundant)&quot;</span><span class="p">)</span><span class="w"> </span><span class="o">+</span>
<span class="w">  </span><span class="nf">ylab</span><span class="p">(</span><span class="s">&quot;Cell count (log10 scale)&quot;</span><span class="p">)</span><span class="w"> </span><span class="o">+</span>
<span class="w">  </span><span class="n">ggsci</span><span class="o">::</span><span class="nf">scale_color_locuszoom</span><span class="p">()</span>
</code></pre></div>
<p><img alt="" src="../single_cell_files/figure-markdown/repertoire-clonality-line-1.png" /></p>
</div>
</div>
</div>
<ul>
<li><strong>Possible interpretation:</strong><ul>
<li>If you see a sharp drop ("hockey stick" curve), a handful of
    clones dominate. This is often a hallmark of strong
    antigen-driven selection.</li>
<li>Flatter curves are typically seen in healthy tissues or naive
    compartments, reflecting higher diversity.</li>
</ul>
</li>
</ul>
<h4 id="2-clonal-space-partitioning">2. Clonal space partitioning</h4>
<p>Sometimes, you want to summarize clonal expansion in a single barplot
splitting each repertoire into bins by abundance. For example:</p>
<ul>
<li>"Hyperexpanded": clones with ≥1% of all cells</li>
<li>"Large": ≥0.1%, etc.</li>
<li>"Rare": all the way down to singletons</li>
</ul>
<p>This lets you instantly compare, for example, how "oligoclonal" each
tissue or cluster is. In other words, this helps you explore how much
"hyperexpanded" is.</p>
<div class="tabbed-set tabbed-alternate" data-tabs="29:1"><input checked="checked" id="__tabbed_29_1" name="__tabbed_29" type="radio" /><div class="tabbed-labels"><label for="__tabbed_29_1">R</label></div>
<div class="tabbed-content">
<div class="tabbed-block">
<div class="cell-code highlight"><pre><span></span><code><span class="n">clonal_prop</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">airr_clonality_prop</span><span class="p">(</span><span class="n">idata</span><span class="p">)</span>

<span class="c1"># Plot: what fraction of the repertoire falls into each clonal bin</span>
<span class="nf">ggplot</span><span class="p">(</span><span class="n">clonal_prop</span><span class="p">,</span><span class="w"> </span><span class="nf">aes</span><span class="p">(</span><span class="n">x</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">Tissue</span><span class="p">,</span><span class="w"> </span><span class="n">y</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">occupied_prop</span><span class="p">,</span><span class="w"> </span><span class="n">fill</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">clonal_prop_bin</span><span class="p">))</span><span class="w"> </span><span class="o">+</span>
<span class="w">  </span><span class="nf">geom_col</span><span class="p">(</span><span class="n">position</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;fill&quot;</span><span class="p">)</span><span class="w"> </span><span class="o">+</span>
<span class="w">  </span><span class="nf">ggtitle</span><span class="p">(</span><span class="s">&quot;Fraction of repertoire: hyperexpanded, large, rare clones&quot;</span><span class="p">)</span><span class="w"> </span><span class="o">+</span>
<span class="w">  </span><span class="nf">ylab</span><span class="p">(</span><span class="s">&quot;Fraction of cells&quot;</span><span class="p">)</span><span class="w"> </span><span class="o">+</span>
<span class="w">  </span><span class="nf">xlab</span><span class="p">(</span><span class="s">&quot;Tissue&quot;</span><span class="p">)</span><span class="w"> </span><span class="o">+</span>
<span class="w">  </span><span class="n">ggsci</span><span class="o">::</span><span class="nf">scale_fill_locuszoom</span><span class="p">()</span><span class="w"> </span><span class="o">+</span>
<span class="w">  </span><span class="nf">theme</span><span class="p">(</span><span class="n">axis.text.x</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">element_text</span><span class="p">(</span><span class="n">angle</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">45</span><span class="p">,</span><span class="w"> </span><span class="n">hjust</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">1</span><span class="p">,</span><span class="w"> </span><span class="n">vjust</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">1</span><span class="p">))</span>
</code></pre></div>
<p><img alt="" src="../single_cell_files/figure-markdown/repertoire-clonality-prop-1.png" /></p>
</div>
</div>
</div>
<ul>
<li><strong>Possible interpretation:</strong><ul>
<li>A tumor-infiltrated sample often shows a big block of
    "hyperexpanded" (or "large") clones, reflecting a focused immune
    response or selection.</li>
<li>Healthy tissues or controls should have more "rare" or "small"
    clones---indicating diverse, unperturbed immunity.</li>
</ul>
</li>
</ul>
<p>You can repeat this for cell clusters, different patients, or
experimental groups simply by re-aggregating repertoires (see earlier
sections).</p>
<h4 id="3-clonal-rank-bins">3. Clonal rank bins</h4>
<p>Useful for fine-grained analysis: how much of the repertoire is
explained by the top 10, top 100, or top 1000 clones?</p>
<div class="tabbed-set tabbed-alternate" data-tabs="30:1"><input checked="checked" id="__tabbed_30_1" name="__tabbed_30" type="radio" /><div class="tabbed-labels"><label for="__tabbed_30_1">R</label></div>
<div class="tabbed-content">
<div class="tabbed-block">
<div class="cell-code highlight"><pre><span></span><code><span class="n">clonal_rank</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">airr_clonality_rank</span><span class="p">(</span>
<span class="w">  </span><span class="n">idata</span><span class="p">,</span>
<span class="w">  </span><span class="n">bins</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="m">10</span><span class="p">,</span><span class="w"> </span><span class="m">100</span><span class="p">,</span><span class="w"> </span><span class="m">1000</span><span class="p">)</span>
<span class="p">)</span>

<span class="nf">ggplot</span><span class="p">(</span><span class="n">clonal_rank</span><span class="p">,</span><span class="w"> </span><span class="nf">aes</span><span class="p">(</span><span class="n">x</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">Tissue</span><span class="p">,</span><span class="w"> </span><span class="n">y</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">occupied_prop</span><span class="p">,</span><span class="w"> </span><span class="n">fill</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">as.factor</span><span class="p">(</span><span class="n">clonal_rank_bin</span><span class="p">)))</span><span class="w"> </span><span class="o">+</span>
<span class="w">  </span><span class="nf">geom_col</span><span class="p">(</span><span class="n">position</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;fill&quot;</span><span class="p">)</span><span class="w"> </span><span class="o">+</span>
<span class="w">  </span><span class="nf">ggtitle</span><span class="p">(</span><span class="s">&quot;Fraction of repertoire: top-10, top-100, top-1000&quot;</span><span class="p">)</span><span class="w"> </span><span class="o">+</span>
<span class="w">  </span><span class="nf">ylab</span><span class="p">(</span><span class="s">&quot;Fraction of cells&quot;</span><span class="p">)</span><span class="w"> </span><span class="o">+</span>
<span class="w">  </span><span class="nf">xlab</span><span class="p">(</span><span class="s">&quot;Tissue&quot;</span><span class="p">)</span><span class="w"> </span><span class="o">+</span>
<span class="w">  </span><span class="n">ggsci</span><span class="o">::</span><span class="nf">scale_fill_locuszoom</span><span class="p">()</span><span class="w"> </span><span class="o">+</span>
<span class="w">  </span><span class="nf">theme</span><span class="p">(</span><span class="n">axis.text.x</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">element_text</span><span class="p">(</span><span class="n">angle</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">45</span><span class="p">,</span><span class="w"> </span><span class="n">hjust</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">1</span><span class="p">,</span><span class="w"> </span><span class="n">vjust</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">1</span><span class="p">))</span>
</code></pre></div>
<p><img alt="" src="../single_cell_files/figure-markdown/repertoire-clonality-rank-1.png" /></p>
</div>
</div>
</div>
<blockquote>
<p><strong>Tip:</strong> You can also plot this as a heatmap, line plot, or grouped
barplot as needed.</p>
</blockquote>
<ul>
<li>
<p><strong>Best practices &amp; notes:</strong></p>
<ul>
<li>Always compare clonality to other diversity measures (Shannon,
    Simpson, etc.) for a complete picture.</li>
<li>Technical artifacts (PCR/UMI bias) can sometimes mimic true
    expansion---use biological controls and inspect barcodes/cell
    counts per clone if unsure.</li>
<li>When using clusters or annotations, make sure you have enough
    cells per group for the statistics to be meaningful.</li>
</ul>
</li>
</ul>
<h3 id="diversity-how-rich-and-even-are-your-repertoires">Diversity: how rich and even are your repertoires?</h3>
<p>True immune diversity is about more than just the number of unique
sequences. It's about the <em>balance</em> of clone sizes, the <em>richness</em> of
the repertoire, and whether a handful of clones dominate or if the
population is broad and even. Measuring diversity helps you:</p>
<ul>
<li>Compare immune health or perturbation across tissues, patients, or
    clusters</li>
<li>Spot narrowing (loss of diversity) in disease, aging, or after
    therapy</li>
<li>Select features for machine learning and biomarker discovery</li>
</ul>
<p>Diversity is the flip side of clonality:</p>
<ul>
<li><strong>Clonality</strong> highlights <em>enrichment</em> --- how much your repertoire
    is dominated by overexpanded ("clonal") receptors.</li>
<li><strong>Diversity</strong> measures the <em>overall heterogeneity</em> and <em>evenness</em> of
    the repertoire --- are many clones present, and are they well
    balanced, or is the repertoire narrow?</li>
</ul>
<p>In practical terms:</p>
<ul>
<li><strong>High diversity</strong> = many unique, somehow similarly sized clones;
    typical of healthy, naive, or unperturbed immunity.</li>
<li><strong>Low diversity</strong> = few clones dominate (high clonality); seen in
    strong immune responses, cancer, infection, or technical issues.</li>
</ul>
<p><code>immunarch</code> provides multiple diversity estimators, each with its own
interpretation. To see all options or details, run <code>?airr_diversity</code> in
your R console. Supported methods are:</p>
<ol>
<li>
<p><strong><code>airr_diversity_dxx()</code></strong> <em>Coverage diversity</em> --- D50, D20, D80...
    "How many top clones do you need to cover X% of all cells?" Small
    D50 = a few clones dominate (low diversity); large D50 = highly
    polyclonal.</p>
</li>
<li>
<p><strong><code>airr_diversity_chao1()</code></strong> <em>A nonparameteric asymptotic estimator</em>
    of species richness (number of species in a population). One of the
    most used methods for estimating immune repertoire diversity.</p>
</li>
<li>
<p><strong><code>airr_diversity_shannon()</code></strong> <em>Shannon entropy</em> --- a gold
    standard, evenness-aware metric for diversity. Higher values = more
    diverse and even repertoire.</p>
</li>
<li>
<p><strong><code>airr_diversity_pielou()</code></strong> <em>Pielou's evenness</em> --- entropy
    normalized by clone count, ranges [0, 1]. Use when comparing
    repertoires of very different sizes.</p>
</li>
<li>
<p><strong><code>airr_diversity_index()</code></strong> <em>Hill number (q=1)</em> --- the "effective
    number" of clones, combines richness and evenness. Intuitive and
    robust: "How many equally frequent clones would give this entropy?"</p>
</li>
<li>
<p><strong><code>airr_diversity_hill()</code></strong> <em>Hill diversity profile (q=0,1,2, ...)</em>
    Powerful for exploring sensitivity to rare vs. abundant clones.</p>
<ul>
<li>q=0: just unique count ("richness")</li>
<li>q=1: entropy-based (exp(Shannon))</li>
<li>q=2: Simpson's index (abundant clones dominate)</li>
</ul>
</li>
</ol>
<div class="tabbed-set tabbed-alternate" data-tabs="31:1"><input checked="checked" id="__tabbed_31_1" name="__tabbed_31" type="radio" /><div class="tabbed-labels"><label for="__tabbed_31_1">R</label></div>
<div class="tabbed-content">
<div class="tabbed-block">
<div class="cell-code highlight"><pre><span></span><code><span class="c1"># 1. Coverage diversity — D50: how many clones cover 50% of the repertoire?</span>
<span class="n">d50</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">airr_diversity_dxx</span><span class="p">(</span><span class="n">idata</span><span class="p">,</span><span class="w"> </span><span class="n">perc</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">50</span><span class="p">)</span>
<span class="n">p1</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">d50</span><span class="w"> </span><span class="o">|&gt;</span><span class="w"> </span><span class="nf">vis</span><span class="p">()</span>

<span class="c1"># 3. Chao1 — non-parametric estimator of species richness (number of unique receptors)</span>
<span class="n">chao</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">airr_diversity_chao1</span><span class="p">(</span><span class="n">idata</span><span class="p">)</span>
<span class="n">p2</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">chao</span><span class="w"> </span><span class="o">|&gt;</span><span class="w"> </span><span class="nf">vis</span><span class="p">(</span><span class="n">fill</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;Tissue&quot;</span><span class="p">)</span>

<span class="c1"># 3. Shannon entropy — classical diversity (higher = more even)</span>
<span class="n">shannon</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">airr_diversity_shannon</span><span class="p">(</span><span class="n">idata</span><span class="p">)</span>
<span class="n">p3</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">shannon</span><span class="w"> </span><span class="o">|&gt;</span><span class="w"> </span><span class="nf">vis</span><span class="p">(</span><span class="n">xval</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;Cluster&quot;</span><span class="p">)</span>

<span class="c1"># 4. Pielou&#39;s evenness (normalized entropy, 0 = dominated, 1 = perfectly even)</span>
<span class="n">pielou</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">airr_diversity_pielou</span><span class="p">(</span><span class="n">idata</span><span class="p">)</span>
<span class="n">p4</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">pielou</span><span class="w"> </span><span class="o">|&gt;</span><span class="w"> </span><span class="nf">vis</span><span class="p">(</span><span class="n">xval</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;Tissue&quot;</span><span class="p">)</span>
<span class="n">p5</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">pielou</span><span class="w"> </span><span class="o">|&gt;</span><span class="w"> </span><span class="nf">vis</span><span class="p">(</span><span class="n">facet</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;Tissue&quot;</span><span class="p">,</span><span class="w"> </span><span class="n">fill</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;Cluster&quot;</span><span class="p">,</span><span class="w"> </span><span class="n">dir</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;v&quot;</span><span class="p">)</span>

<span class="c1"># 5. Hill number (q=1): the “effective number of clones”</span>
<span class="n">hill1</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">airr_diversity_index</span><span class="p">(</span><span class="n">idata</span><span class="p">)</span>

<span class="c1"># 6. Hill diversity profile (q=0, 1, 2, ...)</span>
<span class="n">hill_profile</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">airr_diversity_hill</span><span class="p">(</span><span class="n">idata</span><span class="p">,</span><span class="w"> </span><span class="n">q</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="m">0</span><span class="p">,</span><span class="w"> </span><span class="m">1</span><span class="p">,</span><span class="w"> </span><span class="m">2</span><span class="p">))</span>

<span class="n">p1</span>
</code></pre></div>
<p><img alt="" src="../single_cell_files/figure-markdown/repertoire-diversity-metrics-1.png" /></p>
<div class="cell-code highlight"><pre><span></span><code><span class="n">p2</span>
</code></pre></div>
<p><img alt="" src="../single_cell_files/figure-markdown/repertoire-diversity-metrics-2.png" /></p>
<div class="cell-code highlight"><pre><span></span><code><span class="n">p3</span>
</code></pre></div>
<p><img alt="" src="../single_cell_files/figure-markdown/repertoire-diversity-metrics-3.png" /></p>
<div class="cell-code highlight"><pre><span></span><code><span class="n">p4</span>
</code></pre></div>
<p><img alt="" src="../single_cell_files/figure-markdown/repertoire-diversity-metrics-4.png" /></p>
<div class="cell-code highlight"><pre><span></span><code><span class="n">p5</span>
</code></pre></div>
<p><img alt="" src="../single_cell_files/figure-markdown/repertoire-diversity-metrics-5.png" /></p>
</div>
</div>
</div>
<h3 id="public-receptor-indices-quantifying-repertoire-overlap">Public receptor indices: quantifying repertoire overlap</h3>
<p>In immunology, public receptors are TCRs or BCRs that appear in multiple
individuals or samples, suggesting convergent responses to shared
antigens, or technical "publicity" due to biases. Measuring the overlap
between repertoires helps you:</p>
<ul>
<li>Find biologically meaningful "public" clonotypes</li>
<li>QC for sample swap, cross-contamination, or replicate similarity</li>
<li>Compare cohort similarity or donor-sharing at scale</li>
</ul>
<p><code>immunarch</code> provides fast tools for quantifying this:</p>
<ul>
<li>
<p><strong><code>airr_public_intersection()</code></strong> <em>Count of shared unique receptors
    between each repertoire pair.</em> Use for overlap heatmaps, or to check
    how much two samples have in common.</p>
</li>
<li>
<p><strong><code>airr_public_jaccard()</code></strong> <em>Jaccard similarity of receptor sets
    between repertoires</em> (⁠|A∩B| / |A∪B|⁠). This normalizes for
    differences in sample size---great for comparing across experiments
    or donors.</p>
</li>
</ul>
<p>For full options and details, run <code>?airr_public</code> in your R console.</p>
<p>Let's compute overlap and Jaccard similarity between all pairs of
repertoires (e.g., samples, tissues, clusters):</p>
<div class="tabbed-set tabbed-alternate" data-tabs="32:1"><input checked="checked" id="__tabbed_32_1" name="__tabbed_32" type="radio" /><div class="tabbed-labels"><label for="__tabbed_32_1">R</label></div>
<div class="tabbed-content">
<div class="tabbed-block">
<div class="cell-code highlight"><pre><span></span><code><span class="c1"># 1. Count of shared receptors: intersection matrix</span>
<span class="n">m_pub</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">airr_public_intersection</span><span class="p">(</span><span class="n">idata</span><span class="p">)</span>

<span class="c1"># 2. Jaccard similarity: size-normalized matrix</span>
<span class="n">m_jac</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">airr_public_jaccard</span><span class="p">(</span><span class="n">idata</span><span class="p">)</span>

<span class="n">m_pub</span><span class="w"> </span><span class="o">|&gt;</span><span class="w"> </span><span class="nf">vis</span><span class="p">()</span>
</code></pre></div>
<p><img alt="" src="../single_cell_files/figure-markdown/repertoire-public-overlap-1.png" /></p>
<div class="cell-code highlight"><pre><span></span><code><span class="n">m_jac</span><span class="w"> </span><span class="o">|&gt;</span><span class="w"> </span><span class="nf">vis</span><span class="p">()</span>
</code></pre></div>
<p><img alt="" src="../single_cell_files/figure-markdown/repertoire-public-overlap-2.png" /></p>
</div>
</div>
</div>
<p><strong>Possible interpretation:</strong></p>
<ul>
<li>
<p><strong>Diagonal</strong>: total unique receptors per repertoire ("richness").</p>
</li>
<li>
<p><strong>Off-diagonal</strong>: degree of sharing.</p>
<ul>
<li>High off-diagonal = strong overlap (potentially "public" clones,
    or replicates).</li>
<li>Low off-diagonal = private repertoires (most unique to each
    sample).</li>
</ul>
</li>
</ul>
<p>Use Jaccard for size-normalized comparison, especially across different
tissues, donors, or experiments.</p>
<blockquote>
<p>We found that there are some overlapped receptors. But what are those
public receptors? The next section will help us extract them from the
data and analyse.</p>
</blockquote>
<hr />
<h2 id="discover-and-annotate-immune-receptors">Discover and annotate immune receptors</h2>
<blockquote>
<p>Please mind that that sections are below are still work in progress.</p>
</blockquote>
<p>All <code>airr_*</code> functions work on the repertoire-level and return
repertoire-level statistics.</p>
<p>All <code>receptor_*</code> functions work on the receptor-level and return
filtered <code>ImmunData</code>-s.</p>
<p>All <code>annotate_*</code> functions work on the receptor-level and return
annotated <code>ImmunData</code>-s, preserving the input data and adding additional
columns for annotations.</p>
<h3 id="annotating-receptors-by-clonality">Annotating receptors by clonality</h3>
<blockquote>
<p>This subsection was composed rather quickly to cover a
highly-requested feature. More polished version will be available on
the next release</p>
</blockquote>
<p>Clonality tells us how overabundant a receptor is inside a repertoire.
Here we add a label per receptor, then pass a single label per cell to
Seurat for easy plotting on UMAP.</p>
<p>You can choose one of two simple rules:</p>
<ul>
<li><strong>By proportion</strong> (recommended to start): bins like <em>Hyperexpanded</em>,
    <em>Large</em>, ... based on the receptor's <code>proportion</code> within the
    repertoire.</li>
<li><strong>By rank</strong>: bins by the receptor's <strong>rank</strong> (top 10, top 100, ...)
    within the repertoire.</li>
</ul>
<p>Both helpers keep your data intact and add new columns to <code>ImmunData</code>.</p>
<blockquote>
<p><strong>Note:</strong> Thresholds are heuristics. They are good defaults, not
universal truths. Adjust them to your study.</p>
</blockquote>
<div class="tabbed-set tabbed-alternate" data-tabs="33:1"><input checked="checked" id="__tabbed_33_1" name="__tabbed_33" type="radio" /><div class="tabbed-labels"><label for="__tabbed_33_1">R</label></div>
<div class="tabbed-content">
<div class="tabbed-block">
<div class="cell-code highlight"><pre><span></span><code><span class="n">idata</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">annotate_clonality_prop</span><span class="p">(</span><span class="n">idata</span><span class="p">)</span>

<span class="n">sdata</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">annotate_seurat</span><span class="p">(</span><span class="n">idata</span><span class="p">,</span><span class="w"> </span><span class="n">sdata</span><span class="p">,</span><span class="w"> </span><span class="n">cols</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;clonal_prop_bin&quot;</span><span class="p">)</span>

<span class="n">Seurat</span><span class="o">::</span><span class="nf">DimPlot</span><span class="p">(</span><span class="n">sdata</span><span class="p">,</span><span class="w"> </span><span class="n">reduction</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;umap&quot;</span><span class="p">,</span><span class="w"> </span><span class="n">group.by</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;clonal_prop_bin&quot;</span><span class="p">,</span><span class="w"> </span><span class="n">shuffle</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kc">TRUE</span><span class="p">)</span>
</code></pre></div>
<p><img alt="" src="../single_cell_files/figure-markdown/receptor-annotate-clonality-1.png" /></p>
<div class="cell-code highlight"><pre><span></span><code><span class="n">idata</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">annotate_clonality_rank</span><span class="p">(</span><span class="n">idata</span><span class="p">)</span>

<span class="n">sdata</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">annotate_seurat</span><span class="p">(</span><span class="n">idata</span><span class="p">,</span><span class="w"> </span><span class="n">sdata</span><span class="p">,</span><span class="w"> </span><span class="n">cols</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;clonal_rank_bin&quot;</span><span class="p">)</span>

<span class="n">Seurat</span><span class="o">::</span><span class="nf">DimPlot</span><span class="p">(</span><span class="n">sdata</span><span class="p">,</span><span class="w"> </span><span class="n">reduction</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;umap&quot;</span><span class="p">,</span><span class="w"> </span><span class="n">group.by</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;clonal_rank_bin&quot;</span><span class="p">,</span><span class="w"> </span><span class="n">shuffle</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kc">TRUE</span><span class="p">)</span>
</code></pre></div>
<p><img alt="" src="../single_cell_files/figure-markdown/receptor-annotate-clonality-2.png" /></p>
</div>
</div>
</div>
<hr />
<h3 id="other-sections">Other sections</h3>
<p>The following sections are under construction as I currently develop the
receptor-level analysis functionality. I plan to release the most
commonly used functions in November 2025. Monitor my
<a href="https://www.linkedin.com/in/vdnaz">LinkedIn</a> and
<a href="https://github.com/immunomind/immunarch/issues/432">GitHub</a> for news
and updates.</p>
<p>The planned topics are:</p>
<ul>
<li>
<p>use external databases to find CMV-specific receptors in the input
    repertoires</p>
</li>
<li>
<p>track specific CMV-related receptors through different tissues and
    highlight them on single-cell UMAP</p>
</li>
<li>
<p>analyse and plot counts of specific receptors vs. expression of
    genes of interest</p>
</li>
<li>
<p>public repertoire analysis - discover receptors which are
    suspiciously overabundant in one groups of repertoires in contrast
    to another groups of repertoires (e.g., control vs condition to
    search for TIL or antigen-specific receptors)</p>
</li>
<li>
<p>all of above but using levenshtein-based similarity instead of exact
    matches</p>
</li>
</ul>
<hr />
<h2 id="wip-analyse-subgroups-of-receptors-of-interest">[WIP] Analyse subgroups of receptors of interest</h2>
<p>This is section is too under constructions. Planned content:</p>
<ul>
<li>
<p>discover receptors of interest, e.g., CMV-specific</p>
</li>
<li>
<p>run analysis to compare sum/avg abundance of such receptors between
    groups to measure the level of signal</p>
</li>
</ul>
<hr />
<h2 id="wip-conclusion">[WIP] Conclusion</h2>
<p>No spoilers. :-)</p>
<hr />
<h2 id="testing-zone">Testing zone</h2>
<p>Small test here and there, don't mind me, nothing suspicious is going on
here...</p>
<div class="tabbed-set tabbed-alternate" data-tabs="34:2"><input checked="checked" id="__tabbed_34_1" name="__tabbed_34" type="radio" /><input id="__tabbed_34_2" name="__tabbed_34" type="radio" /><div class="tabbed-labels"><label for="__tabbed_34_1">R</label><label for="__tabbed_34_2">Python</label></div>
<div class="tabbed-content">
<div class="tabbed-block">
<div class="cell-code highlight"><pre><span></span><code><span class="nf">c</span><span class="p">(</span><span class="m">1</span><span class="p">,</span><span class="m">2</span><span class="p">,</span><span class="m">3</span><span class="p">)</span><span class="w"> </span><span class="o">|&gt;</span><span class="w"> </span><span class="nf">sum</span><span class="p">()</span>
</code></pre></div>
</div>
<div class="tabbed-block">
<div class="cell-code highlight"><pre><span></span><code><span class="nb">max</span><span class="p">([</span><span class="mi">1</span><span class="p">,</span><span class="mi">2</span><span class="p">,</span><span class="mi">3</span><span class="p">])</span>
</code></pre></div>
</div>
</div>
</div>
<p>Text block as a divider...</p>
<div class="tabbed-set tabbed-alternate" data-tabs="35:2"><input checked="checked" id="__tabbed_35_1" name="__tabbed_35" type="radio" /><input id="__tabbed_35_2" name="__tabbed_35" type="radio" /><div class="tabbed-labels"><label for="__tabbed_35_1">R</label><label for="__tabbed_35_2">Python</label></div>
<div class="tabbed-content">
<div class="tabbed-block">
<div class="cell-code highlight"><pre><span></span><code><span class="nf">c</span><span class="p">(</span><span class="m">1</span><span class="p">,</span><span class="m">2</span><span class="p">,</span><span class="m">3</span><span class="p">)</span><span class="w"> </span><span class="o">|&gt;</span><span class="w"> </span><span class="nf">sum</span><span class="p">()</span>
</code></pre></div>
<div class="highlight"><pre><span></span><code>[1] 6
</code></pre></div>
</div>
<div class="tabbed-block">
<div class="cell-code highlight"><pre><span></span><code><span class="nb">max</span><span class="p">([</span><span class="mi">1</span><span class="p">,</span><span class="mi">2</span><span class="p">,</span><span class="mi">3</span><span class="p">])</span>
</code></pre></div>
<div class="highlight"><pre><span></span><code>3
</code></pre></div>
</div>
</div>
</div>







  
  



  




                
              </article>
            </div>
          
          
  <script>var tabs=__md_get("__tabs");if(Array.isArray(tabs))e:for(var set of document.querySelectorAll(".tabbed-set")){var labels=set.querySelector(".tabbed-labels");for(var tab of tabs)for(var label of labels.getElementsByTagName("label"))if(label.innerText.trim()===tab){var input=document.getElementById(label.htmlFor);input.checked=!0;continue e}}</script>

<script>var target=document.getElementById(location.hash.slice(1));target&&target.name&&(target.checked=target.name.startsWith("__tabbed_"))</script>
        </div>
        
      </main>
      
        <footer class="md-footer">
  
  <div class="md-footer-meta md-typeset">
    <div class="md-footer-meta__inner md-grid">
      <div class="md-copyright">
  
  
    Made with
    <a href="https://squidfunk.github.io/mkdocs-material/" target="_blank" rel="noopener">
      Material for MkDocs
    </a>
  
</div>
      
    </div>
  </div>
</footer>
      
    </div>
    <div class="md-dialog" data-md-component="dialog">
      <div class="md-dialog__inner md-typeset"></div>
    </div>
    
    
    
      
      <script id="__config" type="application/json">{"base": "../..", "features": ["navigation.tabs", "navigation.tabs.sticky", "content.tabs.link", "content.code.copy", "navigation.sections", "navigation.expand"], "search": "../../assets/javascripts/workers/search.973d3a69.min.js", "tags": null, "translations": {"clipboard.copied": "Copied to clipboard", "clipboard.copy": "Copy to clipboard", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.placeholder": "Type to start searching", "search.result.term.missing": "Missing", "select.version": "Select version"}, "version": null}</script>
    
    
      <script src="../../assets/javascripts/bundle.f55a23d4.min.js"></script>
      
    
  </body>
</html>